{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# default_exp chem"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Chem\n",
    "\n",
    "> Core comp chem functions implemented with RDKit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "from nbdev.showdoc import *\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/dmai/miniconda3/envs/mrl/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: to-Python converter for boost::shared_ptr<RDKit::FilterCatalogEntry const> already registered; second conversion method ignored.\n",
      "  return f(*args, **kwds)\n"
     ]
    }
   ],
   "source": [
    "# export\n",
    "from mrl.imports import *\n",
    "from mrl.core import *\n",
    "from rdkit import Chem, DataStructs\n",
    "from rdkit.Chem import AllChem, rdMolDescriptors, Descriptors, rdMMPA, QED, RDConfig\n",
    "from rdkit.Chem.Lipinski import RotatableBondSmarts\n",
    "from rdkit.Chem.FilterCatalog import *\n",
    "from rdkit.Chem.Scaffolds import MurckoScaffold\n",
    "try:\n",
    "    # doesnt work for github CI\n",
    "    sys.path.append(os.path.join(RDConfig.RDContribDir, 'SA_Score'))\n",
    "    import sascorer\n",
    "except:\n",
    "    warnings.warn(\"SA score not imported\", Warning)\n",
    "    \n",
    "from rdkit import RDLogger\n",
    "RDLogger.DisableLog('rdApp.*')\n",
    "import selfies as sf\n",
    "import networkx as nx"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RDKit i/o\n",
    "\n",
    "`to_mol` and `to_smile` are functions that make it easy to work with both SMILES strings and RDKit `Mol` objects. For example, if a function requires a `mol` input, adding `mol = to_mol(mol)` allows the function to take either SMILES strings or RDKit mols as input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "def to_mol(smile_or_mol):\n",
    "    if (type(smile_or_mol) == str) or (type(smile_or_mol) == np.str_):\n",
    "        mol = Chem.MolFromSmiles(smile_or_mol)\n",
    "        if mol is not None:\n",
    "            try:\n",
    "                Chem.SanitizeMol(mol)\n",
    "            except:\n",
    "                mol = None\n",
    "    else:\n",
    "        mol = smile_or_mol\n",
    "        \n",
    "    return mol\n",
    "\n",
    "def smart_to_mol(smile_or_mol):\n",
    "    if (type(smile_or_mol) == str) or (type(smile_or_mol) == np.str_):\n",
    "        mol = Chem.MolFromSmarts(smile_or_mol)\n",
    "        if mol is not None:\n",
    "            try:\n",
    "                Chem.SanitizeMol(mol)\n",
    "            except:\n",
    "                mol = None\n",
    "    else:\n",
    "        mol = smile_or_mol\n",
    "        \n",
    "    return mol\n",
    "\n",
    "def to_smile(smile_or_mol):\n",
    "    \n",
    "    if type(smile_or_mol)==Chem.Mol:\n",
    "        smile = Chem.MolToSmiles(smile_or_mol)\n",
    "    else:\n",
    "        smile = smile_or_mol\n",
    "        \n",
    "    return smile\n",
    "\n",
    "def to_kekule(smile_or_mol):\n",
    "    \n",
    "    if type(smile_or_mol)==Chem.Mol:\n",
    "        smile = Chem.MolToSmiles(smile_or_mol, kekuleSmiles=True)\n",
    "    else:\n",
    "        mol = to_mol(smile_or_mol)\n",
    "        Chem.Kekulize(mol)\n",
    "        smile = Chem.MolToSmiles(mol, kekuleSmiles=True)\n",
    "        \n",
    "    return smile\n",
    "    \n",
    "def to_smart(smart_or_mol):\n",
    "    \n",
    "    if type(smart_or_mol)==Chem.Mol:\n",
    "        smart = Chem.MolToSmarts(smart_or_mol)\n",
    "    else:\n",
    "        smart = smart_or_mol\n",
    "        \n",
    "    return smart\n",
    "\n",
    "def to_mols(list_of_inputs):\n",
    "    return maybe_parallel(to_mol, list_of_inputs)\n",
    "\n",
    "def to_smiles(list_of_inputs):\n",
    "    return maybe_parallel(to_smile, list_of_inputs)\n",
    "\n",
    "def to_smarts(list_of_inputs):\n",
    "    return maybe_parallel(to_smart, list_of_inputs)\n",
    "\n",
    "def canon_smile(smile):\n",
    "    return Chem.CanonSmiles(smile)\n",
    "\n",
    "def remove_stereo(smile):\n",
    "    if '@' in smile:\n",
    "        mol = to_mol(smile)\n",
    "        Chem.rdmolops.RemoveStereochemistry(mol)\n",
    "        smile = to_smile(mol)\n",
    "    return smile\n",
    "\n",
    "def smile_to_selfie(smile):\n",
    "    return sf.encoder(smile)\n",
    "\n",
    "def selfie_to_smile(selfie):\n",
    "    smile = None\n",
    "    if selfie is not None:\n",
    "        smile = sf.decoder(selfie)\n",
    "    \n",
    "    if smile is not None:\n",
    "        try:\n",
    "            smile = canon_smile(smile)\n",
    "        except:\n",
    "            smile = None\n",
    "    \n",
    "    return smile\n",
    "\n",
    "def split_selfie(selfie):\n",
    "    if selfie is None:\n",
    "        selfie = ''\n",
    "    return list(sf.split_selfies(selfie))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert type(to_smile('CCC')) == str\n",
    "assert type(to_mol('CCC')) == Chem.Mol\n",
    "assert type(to_smile(Chem.MolFromSmiles('CCC'))) == str\n",
    "assert type(to_mol(Chem.MolFromSmiles('CCC'))) == Chem.Mol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "s = 'O=C(NCc1ccc(Br)cc1F)C(=O)NCC1(Cc2ccccc2)CC1'\n",
    "assert selfie_to_smile(smile_to_selfie(s)) == s"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Misc Functions\n",
    "\n",
    "Miscellaneous RDKit related functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "def neutralize_atoms(mol):\n",
    "    \"Neutralize charges, from rdkit.org/docs/Cookbook.html#neutralizing-molecules\"\n",
    "    mol = to_mol(mol)\n",
    "    pattern = Chem.MolFromSmarts(\"[+1!h0!$([*]~[-1,-2,-3,-4]),-1!$([*]~[+1,+2,+3,+4])]\")\n",
    "    at_matches = mol.GetSubstructMatches(pattern)\n",
    "    at_matches_list = [y[0] for y in at_matches]\n",
    "    if len(at_matches_list) > 0:\n",
    "        for at_idx in at_matches_list:\n",
    "            atom = mol.GetAtomWithIdx(at_idx)\n",
    "            chg = atom.GetFormalCharge()\n",
    "            hcount = atom.GetTotalNumHs()\n",
    "            atom.SetFormalCharge(0)\n",
    "            atom.SetNumExplicitHs(hcount - chg)\n",
    "            atom.UpdatePropertyCache()\n",
    "    return mol\n",
    "\n",
    "def initialize_neutralisation_reactions():\n",
    "    'github.com/kaiwang0112006/clusfps/blob/master/third_party_package/RDKit_2015_03_1/Docs/Book/Cookbook.rst'\n",
    "    patts = (\n",
    "        # Imidazoles\n",
    "        ('[n+;H]', 'n'),\n",
    "        # Amines\n",
    "        ('[N+;!H0]', 'N'),\n",
    "        # Carboxylic acids and alcohols\n",
    "        ('[$([O-]);!$([O-][#7])]', 'O'),\n",
    "        # Thiols\n",
    "        ('[S-;X1]', 'S'),\n",
    "        # Sulfonamides\n",
    "        ('[$([N-;X2]S(=O)=O)]', 'N'),\n",
    "        # Enamines\n",
    "        ('[$([N-;X2][C,N]=C)]', 'N'),\n",
    "        # Tetrazoles\n",
    "        ('[n-]', '[nH]'),\n",
    "        # Sulfoxides\n",
    "        ('[$([S-]=O)]', 'S'),\n",
    "        # Amides\n",
    "        ('[$([N-]C=O)]', 'N'),\n",
    "    )\n",
    "    return [(Chem.MolFromSmarts(x), Chem.MolFromSmiles(y, False)) for x, y in patts]\n",
    "\n",
    "def neutralize_charges(mol, reactions=None):\n",
    "    'github.com/kaiwang0112006/clusfps/blob/master/third_party_package/RDKit_2015_03_1/Docs/Book/Cookbook.rst'\n",
    "    mol = to_mol(mol)\n",
    "    \n",
    "    if reactions is None:\n",
    "        reactions = initialize_neutralisation_reactions()\n",
    "        \n",
    "    replaced = False\n",
    "    \n",
    "    for i, (reactant, product) in enumerate(reactions):\n",
    "        while mol.HasSubstructMatch(reactant):\n",
    "            replaced = True\n",
    "            rms = AllChem.ReplaceSubstructs(mol, reactant, product)\n",
    "            mol = rms[0]\n",
    "            \n",
    "    if replaced:\n",
    "        Chem.SanitizeMol(mol)\n",
    "        \n",
    "    return mol\n",
    "\n",
    "def find_bond_groups(mol):\n",
    "    \"\"\"\n",
    "    Find groups of contiguous rotatable bonds and return them sorted by decreasing size\n",
    "    \n",
    "    https://www.rdkit.org/docs/Cookbook.html\n",
    "    \"\"\"\n",
    "    rot_atom_pairs = mol.GetSubstructMatches(RotatableBondSmarts)\n",
    "    rot_bond_set = set([mol.GetBondBetweenAtoms(*ap).GetIdx() for ap in rot_atom_pairs])\n",
    "    rot_bond_groups = []\n",
    "    while (rot_bond_set):\n",
    "        i = rot_bond_set.pop()\n",
    "        connected_bond_set = set([i])\n",
    "        stack = [i]\n",
    "        while (stack):\n",
    "            i = stack.pop()\n",
    "            b = mol.GetBondWithIdx(i)\n",
    "            bonds = []\n",
    "            for a in (b.GetBeginAtom(), b.GetEndAtom()):\n",
    "                bonds.extend([b.GetIdx() for b in a.GetBonds() if (\n",
    "                    (b.GetIdx() in rot_bond_set) and (not (b.GetIdx() in connected_bond_set)))])\n",
    "            connected_bond_set.update(bonds)\n",
    "            stack.extend(bonds)\n",
    "        rot_bond_set.difference_update(connected_bond_set)\n",
    "        rot_bond_groups.append(tuple(connected_bond_set))\n",
    "    return tuple(sorted(rot_bond_groups, reverse = True, key = lambda x: len(x)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "def draw_mols(mols, legends=None, mols_per_row=3, sub_img_size=(300,300)):\n",
    "    \n",
    "    if not is_container(mols):\n",
    "        mols = [mols]\n",
    "        \n",
    "    mols_per_row = min(mols_per_row, len(mols))\n",
    "        \n",
    "    return Chem.Draw.MolsToGridImage(mols, molsPerRow=mols_per_row,\n",
    "                                    subImgSize=sub_img_size, legends=legends)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mol Descriptors\n",
    "\n",
    "These are a bunch of wrappers for standard RDKit functions. The reason for doing this is RDKit functions can't be pickled, which causes all sorts of problems for multiprocessing. This is fixed by creating a wrapper function\n",
    "\n",
    "For example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parallel execution succeeded\n",
      "Parallel execution succeeded\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    output = maybe_parallel(rdMolDescriptors.CalcExactMolWt, [to_mol('CCC')])\n",
    "    print('Parallel execution succeeded')\n",
    "except:\n",
    "    print('parallel execution failed')\n",
    "    \n",
    "def wrapper(mol):\n",
    "    return rdMolDescriptors.CalcExactMolWt(mol)\n",
    "\n",
    "try:\n",
    "    output = maybe_parallel(wrapper, [to_mol('CCC')])\n",
    "    print('Parallel execution succeeded')\n",
    "except:\n",
    "    print('parallel execution failed')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sadly, having a generic wrapper constructor also fails to pickle because such a wrapper requires an RDKit function as input to construct the wrapper, which brings back the pickle problems (see code example below). This leaves us with manually defining wrapper functions for RDKit functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parallel execution succeeded\n"
     ]
    }
   ],
   "source": [
    "def rdkit_wrapper(rdkit_func):\n",
    "    def wrapper(mol):\n",
    "        return rdkit_func(mol)\n",
    "    \n",
    "    return wrapper\n",
    "\n",
    "try:\n",
    "    output = maybe_parallel(rdkit_wrapper(rdMolDescriptors.CalcExactMolWt), [to_mol('CCC')])\n",
    "    print('Parallel execution succeeded')\n",
    "except:\n",
    "    print('parallel execution failed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "def molwt(mol):\n",
    "    'molecular weight'\n",
    "    return rdMolDescriptors.CalcExactMolWt(mol)\n",
    "\n",
    "def hbd(mol):\n",
    "    'number of hydrogen bond donors'\n",
    "    return rdMolDescriptors.CalcNumHBD(mol)\n",
    "\n",
    "def hba(mol):\n",
    "    'number of hydrogen bond acceptors'\n",
    "    return rdMolDescriptors.CalcNumHBA(mol)\n",
    "    \n",
    "def tpsa(mol):\n",
    "    'total polar surface area'\n",
    "    return rdMolDescriptors.CalcTPSA(mol)\n",
    "    \n",
    "def rotbond(mol):\n",
    "    'number of rotatable bonds'\n",
    "    return rdMolDescriptors.CalcNumRotatableBonds(mol)\n",
    "\n",
    "def loose_rotbond(mol):\n",
    "    'number of rotatable bonds, includes things like amides and esters'\n",
    "    return rdMolDescriptors.CalcNumRotatableBonds(mol, False)\n",
    "\n",
    "def rot_chain_length(mol):\n",
    "    'Length of longest contiguous rotatable bond chain'\n",
    "    output = find_bond_groups(mol)\n",
    "    output = len(output[0]) if output else 0\n",
    "    return output\n",
    "    \n",
    "def fsp3(mol):\n",
    "    'fraction sp3 hybridized atoms'\n",
    "    return rdMolDescriptors.CalcFractionCSP3(mol)\n",
    "    \n",
    "def logp(mol):\n",
    "    'logp'\n",
    "    return Descriptors.MolLogP(mol)\n",
    "    \n",
    "def rings(mol):\n",
    "    'number of rings'\n",
    "    return rdMolDescriptors.CalcNumRings(mol)\n",
    "\n",
    "def max_ring_size(mol):\n",
    "    'size of largest ring'\n",
    "    ring_info = mol.GetRingInfo()\n",
    "    return max((len(r) for r in ring_info.AtomRings()), default=0)\n",
    "\n",
    "def min_ring_size(mol):\n",
    "    'size of smallest ring'\n",
    "    ring_info = mol.GetRingInfo()\n",
    "    return min((len(r) for r in ring_info.AtomRings()), default=0)\n",
    "    \n",
    "def heteroatoms(mol):\n",
    "    'number of heteroatoms'\n",
    "    return rdMolDescriptors.CalcNumHeteroatoms(mol)\n",
    "\n",
    "def all_atoms(mol):\n",
    "    'total number of atoms'\n",
    "    mol = Chem.AddHs(mol)\n",
    "    return mol.GetNumAtoms()\n",
    "\n",
    "def heavy_atoms(mol):\n",
    "    'number of heavy atoms'\n",
    "    return mol.GetNumHeavyAtoms()\n",
    "\n",
    "def formal_charge(mol):\n",
    "    'formal charge'\n",
    "    return Chem.rdmolops.GetFormalCharge(mol)\n",
    "\n",
    "def molar_refractivity(mol):\n",
    "    'molar refractivity'\n",
    "    return Descriptors.MolMR(mol)\n",
    "\n",
    "def aromaticrings(mol):\n",
    "    'number of aromatic rings'\n",
    "    return rdMolDescriptors.CalcNumAromaticRings(mol)\n",
    "\n",
    "def qed(mol):\n",
    "    'QED Score doi:10.1038/nchem.1243Â '\n",
    "    return QED.qed(mol)\n",
    "\n",
    "def sa_score(mol):\n",
    "    'Synthetic Accessability score doi.org/10.1186/1758-2946-1-8'\n",
    "    return sascorer.calculateScore(mol)\n",
    "\n",
    "def num_bridgeheads(mol):\n",
    "    'Number of bridgehead atoms'\n",
    "    return rdMolDescriptors.CalcNumBridgeheadAtoms(mol)\n",
    "\n",
    "def num_spiro(mol):\n",
    "    'Number of spiro atoms'\n",
    "    return rdMolDescriptors.CalcNumSpiroAtoms(mol)\n",
    "\n",
    "def chiral_centers(mol):\n",
    "    'Number of chiral centers'\n",
    "    return len(Chem.FindMolChiralCenters(mol))\n",
    "\n",
    "\n",
    "# Modified from https://github.com/bowenliu16/rl_graph_generation\n",
    "def penalized_logp(mol):\n",
    "    \"\"\"\n",
    "    Reward that consists of log p penalized by SA and # long cycles, \n",
    "    as described in (Kusner et al. 2017). Scores are normalized based on the \n",
    "    statistics of 250k_rndm_zinc_drugs_clean.smi dataset \n",
    "    :param mol: rdkit mol object\n",
    "    :return: float\n",
    "    \"\"\"\n",
    "    # normalization constants, statistics from 250k_rndm_zinc_drugs_clean.smi\n",
    "    mol = to_mol(mol)\n",
    "    logP_mean = 2.4570953396190123\n",
    "    logP_std = 1.434324401111988\n",
    "    SA_mean = -3.0525811293166134\n",
    "    SA_std = 0.8335207024513095\n",
    "    cycle_mean = -0.0485696876403053\n",
    "    cycle_std = 0.2860212110245455\n",
    "\n",
    "    log_p = logp(mol)\n",
    "    SA = -sa_score(mol)\n",
    "\n",
    "    # cycle score\n",
    "    cycle_list = nx.cycle_basis(nx.Graph(\n",
    "        Chem.rdmolops.GetAdjacencyMatrix(mol)))\n",
    "    if len(cycle_list) == 0:\n",
    "        cycle_length = 0\n",
    "    else:\n",
    "        cycle_length = max([len(j) for j in cycle_list])\n",
    "    if cycle_length <= 6:\n",
    "        cycle_length = 0\n",
    "    else:\n",
    "        cycle_length = cycle_length - 6\n",
    "    cycle_score = -cycle_length\n",
    "\n",
    "    normalized_log_p = (log_p - logP_mean) / logP_std\n",
    "    normalized_SA = (SA - SA_mean) / SA_std\n",
    "    normalized_cycle = (cycle_score - cycle_mean) / cycle_std\n",
    "\n",
    "    return normalized_log_p + normalized_SA + normalized_cycle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    _ = maybe_parallel(hbd, [to_mol('CCC')])\n",
    "    output = 'success'\n",
    "except:\n",
    "    output = 'fail'\n",
    "    \n",
    "assert output == 'success'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Substructure Matching\n",
    "\n",
    "This class is used for substructure matching an input `Mol` against a list of SMARTS. \n",
    "\n",
    "__Note__: Substructure matching is tricky. Be sure to verify your SMARTS before putting a large number of them into a filter.\n",
    "\n",
    "`CatalogMatch` functions as a base class to match `Mol` objects against any generic catalog. `has_match` will return a single boolean value for if the `Mol` matches one of he filters in the catalog. `get_matches` will return a list of bools for all elements in the catalog. `percent_matches` returns a list of floats for what percentate of filters match.\n",
    "\n",
    "`SMARTSMatch` will generate a catalog from a list of SMARTS\n",
    "\n",
    "`PAINSMatch`, `PAINSAMatch`, `PAINSBMatch` and `PAINSCMatch` specify different PAINS catalogs present in RDKit (see [here](https://www.rdkit.org/docs/source/rdkit.Chem.rdfiltercatalog.html#rdkit.Chem.rdfiltercatalog.FilterCatalogParams.FilterCatalogs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "class Catalog():\n",
    "    '''\n",
    "    Base Class for SMARTS matching\n",
    "    \n",
    "    Inputs: \n",
    "    \n",
    "    - `catalog FilterCatalog`: RDKit `FilterCatalog`\n",
    "    '''\n",
    "    def __init__(self, catalog):\n",
    "        self.catalog = catalog\n",
    "        self.names = [self.catalog.GetEntryWithIdx(i).GetDescription() \n",
    "                      for i in range(self.catalog.GetNumEntries())]\n",
    "            \n",
    "    def __call__(self, mol, criteria=None): \n",
    "        '''\n",
    "        __call__ - run `mol` against `self.catalog`\n",
    "        \n",
    "        Inputs:\n",
    "        \n",
    "        - `mol [Chem.Mol, list[Chem.Mol]]`: input mols\n",
    "        \n",
    "        - `criteria ['any', 'all', float, int]`: match criteria. \n",
    "        (match any filter, match all filters, match float percent of filters, \n",
    "        match int number of filters)\n",
    "        \n",
    "        If `any`, returns `True` if any smarts match. \n",
    "        \n",
    "        If `all`, returns `True` if all smarts match. \n",
    "        \n",
    "        If `float`, returns `True` if more than `float` percent (inclusive) \n",
    "        of smarts match. \n",
    "        \n",
    "        If `int`, returns `True` if more than `int` total (inclusive) smarts match\n",
    "        '''\n",
    "        \n",
    "        if type(mol)==Chem.Mol:\n",
    "            mol = [mol]\n",
    "        \n",
    "        if criteria=='all':\n",
    "            output = [all(i) for i in maybe_parallel(self.get_matches, mol)]\n",
    "        elif criteria=='any':\n",
    "            output = maybe_parallel(self.has_match, mol)\n",
    "        elif type(criteria)==float:\n",
    "            output = [i>=criteria for i in maybe_parallel(self.percent_matches, mol)]\n",
    "        elif type(criteria)==int:\n",
    "            output = [i>=criteria for i in maybe_parallel(self.num_matches, mol)]\n",
    "        else:\n",
    "            output = maybe_parallel(self.get_matches, mol)\n",
    "            \n",
    "        if len(output)==1:\n",
    "            output = output[0]\n",
    "            \n",
    "        return output\n",
    "            \n",
    "    def has_match(self, mol):\n",
    "        mol = to_mol(mol)\n",
    "        return self.catalog.HasMatch(mol)\n",
    "    \n",
    "    def get_matches(self, mol):\n",
    "        mol = to_mol(mol)\n",
    "        matches = [i.GetDescription() for i in self.catalog.GetMatches(mol)]\n",
    "        return [i in matches for i in self.names]\n",
    "    \n",
    "    def num_matches(self, mol):\n",
    "        mol = to_mol(mol)\n",
    "        matches = [i.GetDescription() for i in self.catalog.GetMatches(mol)]\n",
    "        return len(matches)\n",
    "    \n",
    "    def percent_matches(self, mol):\n",
    "        num_matches = self.num_matches(mol)\n",
    "        return num_matches/len(self.names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "<h4 id=\"Catalog.__call__\" class=\"doc_header\"><code>Catalog.__call__</code><a href=\"__main__.py#L16\" class=\"source_link\" style=\"float:right\">[source]</a></h4>\n",
       "\n",
       "> <code>Catalog.__call__</code>(**`mol`**, **`criteria`**=*`None`*)\n",
       "\n",
       "__call__ - run `mol` against `self.catalog`\n",
       "\n",
       "Inputs:\n",
       "\n",
       "- `mol [Chem.Mol, list[Chem.Mol]]`: input mols\n",
       "\n",
       "- `criteria ['any', 'all', float, int]`: match criteria. \n",
       "(match any filter, match all filters, match float percent of filters, \n",
       "match int number of filters)\n",
       "\n",
       "If `any`, returns `True` if any smarts match. \n",
       "\n",
       "If `all`, returns `True` if all smarts match. \n",
       "\n",
       "If `float`, returns `True` if more than `float` percent (inclusive) \n",
       "of smarts match. \n",
       "\n",
       "If `int`, returns `True` if more than `int` total (inclusive) smarts match"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "show_doc(Catalog.__call__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "    \n",
    "class SmartsCatalog(Catalog):\n",
    "    '''\n",
    "    Class for SMARTS matching\n",
    "    \n",
    "    Inputs: `smarts`, list of smarts\n",
    "    '''\n",
    "    def __init__(self, smarts):\n",
    "        self.smarts = smarts\n",
    "        catalog = FilterCatalog()\n",
    "        for s in smarts:\n",
    "            catalog.AddEntry(FilterCatalogEntry(s, SmartsMatcher(s,s)))\n",
    "            \n",
    "        super().__init__(catalog)\n",
    "        \n",
    "    def get_matches(self, mol):\n",
    "        mol = to_mol(mol)\n",
    "        matches = [i.GetDescription() for i in self.catalog.GetMatches(mol)]\n",
    "        return [i in matches for i in self.smarts]\n",
    "        \n",
    "        \n",
    "class ParamsCatalog(Catalog):\n",
    "    '''\n",
    "    Generates `CatalogMatch` object from `params`, a `FilterCatalogParams` object\n",
    "    '''\n",
    "    def __init__(self, params):\n",
    "        p = FilterCatalogParams()\n",
    "        p.AddCatalog(params)\n",
    "        catalog = FilterCatalog(params)\n",
    "        super().__init__(catalog)\n",
    "        \n",
    "class PAINSCatalog(ParamsCatalog):\n",
    "    \"Full PAINS filter matching\"\n",
    "    def __init__(self):\n",
    "        super().__init__(FilterCatalogParams.FilterCatalogs.PAINS)\n",
    "        \n",
    "class PAINSACatalog(ParamsCatalog):\n",
    "    \"PAINS A filter matching\"\n",
    "    def __init__(self):\n",
    "        super().__init__(FilterCatalogParams.FilterCatalogs.PAINS_A)\n",
    "        \n",
    "class PAINSBCatalog(ParamsCatalog):\n",
    "    \"PAINS B filter matching\"\n",
    "    def __init__(self):\n",
    "        super().__init__(FilterCatalogParams.FilterCatalogs.PAINS_B)\n",
    "        \n",
    "class PAINSCCatalog(ParamsCatalog):\n",
    "    \"PAINS C filter matching\"\n",
    "    def __init__(self):\n",
    "        super().__init__(FilterCatalogParams.FilterCatalogs.PAINS_C)\n",
    "\n",
    "class ZINCCatalog(ParamsCatalog):\n",
    "    \"ZINC filter matching\"\n",
    "    def __init__(self):\n",
    "        super().__init__(FilterCatalogParams.FilterCatalogs.ZINC)\n",
    "        \n",
    "class BRENKCatalog(ParamsCatalog):\n",
    "    \"BRENK filter matching\"\n",
    "    def __init__(self):\n",
    "        super().__init__(FilterCatalogParams.FilterCatalogs.BRENK)\n",
    "        \n",
    "class NIHCatalog(ParamsCatalog):\n",
    "    \"NIH filter matching\"\n",
    "    def __init__(self):\n",
    "        super().__init__(FilterCatalogParams.FilterCatalogs.NIH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "smarts = [\n",
    "    '[*]-[#6]1:[#6]:[#6](-[#0]):[#6]:[#6](-[*]):[#6]:1',\n",
    "    '[*]-[#6]1:[#6]:[#6](-[*]):[#6]:[#6]:[#6]:1',\n",
    "    '[*]-[#6]1:[#6]:[#6]:[#6]:[#6]:[#6]:1',\n",
    "    '[*]-[#6]1:[#6]:[#6](-[#7]-[*]):[#6]:[#6]:[#6]:1',\n",
    "    '[#6]1:[#6]:[#7]:[#6]:[#6]:[#6]:1'\n",
    "]\n",
    "\n",
    "sm = SmartsCatalog(smarts)\n",
    "\n",
    "smiles = [\n",
    "    'c1ccccc1',\n",
    "    'Cc1cc(NC)ccc1',\n",
    "    'Cc1cc(NC)cnc1',\n",
    "    'Cc1cccc(NCc2ccccc2)c1'\n",
    "]\n",
    "\n",
    "mols = [to_mol(i) for i in smiles]\n",
    "\n",
    "assert sm(mols, criteria='any') == [False, True, True, True]\n",
    "assert sm(mols, criteria=0.5) == [False, True, False, True]\n",
    "assert sm(mols[1], criteria=3)==True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fingerprints\n",
    "\n",
    "This section deals with creating and manipulating molecular fingerprints. Below are functions for generating different forms of Morgan fingerprints (ECFP4, ECFP6, FCFP4, FCFP6). Fingerprints by default are generated as RDKit `ExplicitBitVect` objects, but can be converted to numpy arrays using the `fp_to_array` function.\n",
    "\n",
    "Fingerprint similarity functions using Tanimoto, Dice and Cosine metrics are implemented for both `ExplicitBitVect` and `ndarray` objects.\n",
    "\n",
    "__Note__ following cheminformatics convention, fingerprint metrics are implemented as _similarities_ rather than _distances_. The metrics used have the relationship `similarity = 1 - distance`. For using fingerprint difference metrics in machine learning applications, be sure you are using the correct relationship (similarity vs difference) for your task."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "def morgan_fp(mol, radius=3, nbits=2048, use_features=False):\n",
    "    \"morgan fingerprint\"\n",
    "    mol = to_mol(mol)\n",
    "    fp = rdMolDescriptors.GetMorganFingerprintAsBitVect(mol, radius, nBits=nbits, useFeatures=use_features)\n",
    "    return fp\n",
    "\n",
    "def ECFP4(mol):\n",
    "    \"ECFP4 Fingerprint\"\n",
    "    return morgan_fp(mol, radius=2, use_features=False)\n",
    "\n",
    "def ECFP6(mol):\n",
    "    \"ECFP6 Fingerprint\"\n",
    "    return morgan_fp(mol, radius=3, use_features=False)\n",
    "\n",
    "def FCFP4(mol):\n",
    "    \"FCFP4 Fingerprint\"\n",
    "    return morgan_fp(mol, radius=2, use_features=True)\n",
    "\n",
    "def FCFP6(mol):\n",
    "    \"FCFP6 Fingerprint\"\n",
    "    return morgan_fp(mol, radius=3, use_features=True)\n",
    "\n",
    "def failsafe_fp(mol, fp_function):\n",
    "    'Returns vector of zeros if failure'\n",
    "    if mol is None:\n",
    "        mol = to_mol('')\n",
    "\n",
    "    try:\n",
    "        output = fp_function(mol)\n",
    "    except:\n",
    "        mol = to_mol('')\n",
    "        output = fp_function(mol)\n",
    "        \n",
    "    return output\n",
    "\n",
    "def fp_to_array(fp):\n",
    "    \"Converts RDKit `ExplicitBitVec` to numpy array\"\n",
    "    \n",
    "    if is_container(fp):\n",
    "        return np.stack([fp_to_array(i) for i in fp])\n",
    "    else:\n",
    "        # note `ConvertToNumpyArray` ~25x faster than `np.asarray`\n",
    "        arr = np.zeros((1,), dtype=np.int8)\n",
    "        DataStructs.ConvertToNumpyArray(fp, arr)\n",
    "\n",
    "        return arr\n",
    "\n",
    "def tanimoto(fps1, fps2):\n",
    "    'Tanimoto similarity'\n",
    "    return 1-distance.cdist(fps1, fps2, metric='jaccard')\n",
    "\n",
    "def tanimoto_rd(fp, fps):\n",
    "    return DataStructs.BulkTanimotoSimilarity(fp, fps)\n",
    "\n",
    "def dice(fps1, fps2):\n",
    "    'Dice similarity'\n",
    "    return 1-distance.cdist(fps1, fps2, metric='dice')\n",
    "\n",
    "def dice_rd(fp, fps):\n",
    "    return DataStructs.BulkDiceSimilarity(fp, fps)\n",
    "\n",
    "def cosine(fps1, fps2):\n",
    "    'Cosine similarity'\n",
    "    return 1-distance.cdist(fps1, fps2, metric='cosine')\n",
    "\n",
    "def cosine_rd(fp, fps):\n",
    "    return DataStructs.BulkCosineSimilarity(fp, fps)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When computing similarities between fingerprints, several things need to be lined up. Different methods are needed for different fingerprint formats (`ndarray` vs `ExplicitBitVect`) and different distance metrics.\n",
    "\n",
    "The `FP` class holds logic to make this easy. \n",
    "\n",
    "The `FP.get_fingerprint` function allows for parallel processing of fingerprint generation.\n",
    "\n",
    "The `FP.fingerprint_similarity` routes fingerprints to the correct similarity function based on the fingerprint's array type and the simillarity metric used.\n",
    "\n",
    "For cases where instantiating a class isn't helpful, `get_fingerprint` and `fingerprint_similarities` work as functional wrappers around `FP`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "class FP():\n",
    "    '''\n",
    "    FP - class for manipulating molecular fingerprints\n",
    "    \n",
    "    `fp = FP()`\n",
    "    \n",
    "    `mol_fp = fp.get_fingerprint(mol, 'ECFP6')`\n",
    "    '''\n",
    "    def __init__(self):\n",
    "        self.similarities = {\n",
    "            'tanimoto' : {'rdkit' : tanimoto_rd,\n",
    "                          'numpy' : tanimoto},\n",
    "            'dice'     : {'rdkit' : dice_rd,\n",
    "                          'numpy' : dice},\n",
    "            'cosine'   : {'rdkit' : cosine_rd,\n",
    "                          'numpy' : dice}\n",
    "        }\n",
    "        \n",
    "        self.fps = {'ECFP4' : ECFP4,\n",
    "                    'ECFP6' : ECFP6,\n",
    "                    'FCFP4' : FCFP4,\n",
    "                    'FCFP6' : FCFP6}\n",
    "        \n",
    "    def get_fingerprint(self, mol, fp_type='ECFP6', output_type='rdkit'):\n",
    "        '''\n",
    "        get_fingerprint - Generates fingerprint for `mol`.\n",
    "        \n",
    "        Inputs:\n",
    "        \n",
    "        - `mol [Chem.Mol, list[Chem.Mol]]` - input mols\n",
    "        \n",
    "        - `fp_type str`: Fingerprint type, must be a key of `FP.fps`\n",
    "        \n",
    "        - `output_type str['rdkit', 'numpy']`: Output datatype. Numpy \n",
    "        `ndarray` or RDKit `ExplicitBitVec`        \n",
    "        '''\n",
    "        method = self.fps[fp_type]\n",
    "\n",
    "        fps = maybe_parallel(method, mol)\n",
    "\n",
    "        if output_type == 'numpy':\n",
    "            fps = fp_to_array(fps)\n",
    "\n",
    "        return fps\n",
    "    \n",
    "    def fingerprint_similarity(self, fps1, fps2, metric):\n",
    "        '''\n",
    "        fingerprint_similarity - Computes the similarity \n",
    "        between `fps1` and `fps2` using `metric`\n",
    "        \n",
    "        Inputs:\n",
    "        \n",
    "        - `fps1 [ndarray, ExplicitBitVect]`: first fingerprint set\n",
    "        \n",
    "        - `fps2 [ndarray, ExplicitBitVect]`: second fingerprint set\n",
    "        \n",
    "        Returns:\n",
    "        \n",
    "        - `similarities [ndarray]`: matrix of similarities between all \n",
    "        fingerprints in `fps1` and all fingerprints in `fps2`\n",
    "        \n",
    "        Fingerprints can either be a Numpy `ndarray` or an RDKit `ExplicitBitVec`. \n",
    "        Both fingerprint inputs must be the same datatype.\n",
    "        \n",
    "        Numpy fingerprints can eiher be a 1D vector or a 2D matrix \n",
    "        of stacked fingerprints\n",
    "        \n",
    "        RDKit fingerprints can either be an `ExplicitBitVec` or a list of \n",
    "        `ExplicitBitVec` objects.\n",
    "        \n",
    "        '''\n",
    "        input_type = self._np_or_rd(fps1)\n",
    "        \n",
    "        metric = self.similarities[metric][input_type]\n",
    "        \n",
    "        if input_type == 'numpy':\n",
    "            fps1 = self._maybe_stack(fps1)\n",
    "            fps2 = self._maybe_stack(fps2)\n",
    "            \n",
    "            similarities = self._fp_similarity_numpy(fps1, fps2, metric)\n",
    "            \n",
    "        else:\n",
    "            fps1 = self._maybe_list(fps1)\n",
    "            fps2 = self._maybe_list(fps2)\n",
    "        \n",
    "            similarities = self._fp_similarity_rdkit(fps1, fps2, metric)\n",
    "            \n",
    "        return similarities\n",
    "    \n",
    "    def _maybe_stack(self, inp):\n",
    "        \"helper function to prepare numpy fingerprints\"\n",
    "        if type(inp)==list:\n",
    "            inp = np.stack(inp)\n",
    "            \n",
    "        if inp.ndim == 1:\n",
    "            inp = inp[None,:]\n",
    "        \n",
    "        return inp\n",
    "    \n",
    "    def _maybe_list(self, inp):\n",
    "        \"helper function to prepare rdkit fingerprints\"\n",
    "        if type(inp) == DataStructs.cDataStructs.ExplicitBitVect:\n",
    "            inp = [inp]\n",
    "            \n",
    "        return inp\n",
    "    \n",
    "    def _fp_similarity_rdkit(self, fps1, fps2, metric):\n",
    "        'FP similarity using RDKit methods'\n",
    "        similarities = []\n",
    "\n",
    "        for fp in fps1:\n",
    "            dist = metric(fp, fps2)\n",
    "            similarities.append(dist)\n",
    "\n",
    "        return np.squeeze(np.array(similarities))\n",
    "    \n",
    "    def _fp_similarity_numpy(self, fps1, fps2, metric):\n",
    "        'FP similarity using Numpy methods'\n",
    "        return np.squeeze(metric(fps1, fps2))\n",
    "    \n",
    "    def _np_or_rd(self, inp):\n",
    "        if type(inp) == np.ndarray:\n",
    "            output = 'numpy'\n",
    "        elif type(inp) == list:\n",
    "            output = self._np_or_rd(inp[0])\n",
    "        else:\n",
    "            output = 'rdkit'\n",
    "\n",
    "        return output\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "<h4 id=\"FP.get_fingerprint\" class=\"doc_header\"><code>FP.get_fingerprint</code><a href=\"__main__.py#L26\" class=\"source_link\" style=\"float:right\">[source]</a></h4>\n",
       "\n",
       "> <code>FP.get_fingerprint</code>(**`mol`**, **`fp_type`**=*`'ECFP6'`*, **`output_type`**=*`'rdkit'`*)\n",
       "\n",
       "get_fingerprint - Generates fingerprint for `mol`.\n",
       "\n",
       "Inputs:\n",
       "\n",
       "- `mol [Chem.Mol, list[Chem.Mol]]` - input mols\n",
       "\n",
       "- `fp_type str`: Fingerprint type, must be a key of [`FP.fps`](/mrl/chem#FP.fps)\n",
       "\n",
       "- `output_type str['rdkit', 'numpy']`: Output datatype. Numpy \n",
       "`ndarray` or RDKit `ExplicitBitVec`        "
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "show_doc(FP.get_fingerprint)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "<h4 id=\"FP.fingerprint_similarity\" class=\"doc_header\"><code>FP.fingerprint_similarity</code><a href=\"__main__.py#L48\" class=\"source_link\" style=\"float:right\">[source]</a></h4>\n",
       "\n",
       "> <code>FP.fingerprint_similarity</code>(**`fps1`**, **`fps2`**, **`metric`**)\n",
       "\n",
       "fingerprint_similarity - Computes the similarity \n",
       "between `fps1` and `fps2` using `metric`\n",
       "\n",
       "Inputs:\n",
       "\n",
       "- `fps1 [ndarray, ExplicitBitVect]`: first fingerprint set\n",
       "\n",
       "- `fps2 [ndarray, ExplicitBitVect]`: second fingerprint set\n",
       "\n",
       "Returns:\n",
       "\n",
       "- `similarities [ndarray]`: matrix of similarities between all \n",
       "fingerprints in `fps1` and all fingerprints in `fps2`\n",
       "\n",
       "Fingerprints can either be a Numpy `ndarray` or an RDKit `ExplicitBitVec`. \n",
       "Both fingerprint inputs must be the same datatype.\n",
       "\n",
       "Numpy fingerprints can eiher be a 1D vector or a 2D matrix \n",
       "of stacked fingerprints\n",
       "\n",
       "RDKit fingerprints can either be an `ExplicitBitVec` or a list of \n",
       "`ExplicitBitVec` objects."
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "show_doc(FP.fingerprint_similarity)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "def get_fingerprint(mol, fp_type, output_type='rdkit'):\n",
    "    fp = FP()\n",
    "    return fp.get_fingerprint(mol, fp_type, output_type=output_type)\n",
    "\n",
    "def fingerprint_similarities(fps1, fps2, metric):\n",
    "    fp = FP()\n",
    "    return fp.fingerprint_similarity(fps1, fps2, metric)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fp = FP()\n",
    "fps = fp.get_fingerprint(mols, fp_type='ECFP4', output_type='rdkit')\n",
    "fps_np = fp_to_array(fps)\n",
    "\n",
    "assert np.allclose(fp.fingerprint_similarity(fps, fps, 'tanimoto'), \n",
    "                   fp.fingerprint_similarity(fps_np, fps_np, 'tanimoto'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Custom Fingerpint Functions\n",
    "\n",
    "Here is an example on how to add new fingerprint functions and distance functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# new fingerprint function\n",
    "\n",
    "def my_fp(mol):\n",
    "    mol = to_mol(mol)\n",
    "    fp =  AllChem.RDKFingerprint(mol)\n",
    "    return fp\n",
    "\n",
    "class MyFP(FP):\n",
    "    def __init__(self):\n",
    "        super().__init__()\n",
    "        self.fps['my_fp'] = my_fp\n",
    "        \n",
    "fp = MyFP()\n",
    "fps = fp.get_fingerprint(mols, fp_type='my_fp', output_type='rdkit')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.00195312, 0.00097656, 0.00048828, 0.00146484],\n",
       "       [0.00097656, 0.01123047, 0.00585938, 0.00830078],\n",
       "       [0.00048828, 0.00585938, 0.01171875, 0.00439453],\n",
       "       [0.00146484, 0.00830078, 0.00439453, 0.01806641]])"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# new distance function\n",
    "\n",
    "def my_dist(fps1, fps2):\n",
    "    # make sure your distance function works on binary/boolean arrays!!\n",
    "    return 1-distance.cdist(fps1, fps2, metric='russellrao')\n",
    "\n",
    "def my_dist_rd(fp, fps):\n",
    "    # make sure the RDKit method gives the same result as scipy, not always the case\n",
    "    return DataStructs.BulkRusselSimilarity(fp, fps)\n",
    "\n",
    "class MyFP(FP):\n",
    "    def __init__(self):\n",
    "        super().__init__()\n",
    "        self.similarities['my_metric'] = {'rdkit' : my_dist_rd,\n",
    "                                          'numpy' : my_dist}\n",
    "                \n",
    "fp = MyFP()\n",
    "fps = fp.get_fingerprint(mols, fp_type='ECFP6', output_type='numpy')\n",
    "fp.fingerprint_similarity(fps, fps, 'my_metric')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mol Operations\n",
    "\n",
    "Functions for editing or manipulating `Mol` objects.\n",
    "\n",
    "Fragmenting functions like `fragment_smile` break molecules into fragments by cutting single bonds.\n",
    "\n",
    "`fuse_on_atom_mapping` fuses fragments following RDKit's atom mapping conventions.\n",
    "\n",
    "`[*:1]-R1-[*:2] + [*:1]-R2 >> [*:2]-R1-R2`\n",
    "\n",
    "`fuse_on_link` relies on user-defined linkages such as heavy atoms.\n",
    "\n",
    "`[Rb]-R1-[Pb] + [Rb]-R2 >> [Pb]-R1-R2`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "def fragment_mol(mol, max_cuts, return_mols=False):\n",
    "    'Wrapper for RDKit mol fragmentation'\n",
    "    return rdMMPA.FragmentMol(mol, maxCuts=max_cuts, resultsAsMols=return_mols)\n",
    "\n",
    "def fragment_smile(smile, cuts):\n",
    "    '''\n",
    "    fragment_smile - fragment `smile` based on `cuts`\n",
    "    \n",
    "    `smile str`: smiles string to fragment\n",
    "    \n",
    "    `cuts [int, list[int]]`: Number of cuts to make. If list, \n",
    "     code iterates over items in `cuts` and generates fragments from each value. \n",
    "     \n",
    "     Note that the RDKit fragmentation uses only a single cut value. So fragmenting with \n",
    "     5 cuts will not include the result of fragmenting with 4 cuts\n",
    "    '''\n",
    "    mol = to_mol(smile)\n",
    "    Chem.rdmolops.RemoveStereochemistry(mol)\n",
    "    \n",
    "    if type(cuts)==int:\n",
    "        frags = fragment_mol(mol, cuts, return_mols=False)\n",
    "        frags = deduplicate_list(flatten_list_of_lists(frags))\n",
    "    else:\n",
    "        frags = []\n",
    "        for cut in cuts:\n",
    "            f = fragment_mol(mol, cut, return_mols=False)\n",
    "            f = deduplicate_list(flatten_list_of_lists(f))\n",
    "            frags += f\n",
    "        \n",
    "        frags = deduplicate_list(frags)\n",
    "\n",
    "    p = re.compile('\\[\\*(.*?)]')\n",
    "    \n",
    "    clean_frags = []\n",
    "    \n",
    "    for frag in frags:\n",
    "        if frag: # case of ''\n",
    "            frag = p.sub('[*]', frag)\n",
    "            frag = Chem.CanonSmiles(frag)\n",
    "            if '.' in frag:\n",
    "                frag = frag.split('.')\n",
    "            else:\n",
    "                frag = [frag]\n",
    "                \n",
    "            frag = [canon_smile(i) for i in frag]\n",
    "            clean_frags += frag\n",
    "            \n",
    "    clean_frags = deduplicate_list(clean_frags)\n",
    "    \n",
    "    return clean_frags\n",
    "\n",
    "def fragment_smiles(smiles, cuts):\n",
    "    'fragment and deduplicate a list of smiles'\n",
    "    frags = maybe_parallel(fragment_smile, smiles, cuts=cuts)\n",
    "    frags = deduplicate_list(flatten_list_of_lists(frags))\n",
    "    return frags"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['*CCCCCC', '*CCCC', '*CCC', '*CCCCCCC', '*CCCCC', '*C', '*CC']"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fragment_smile('CCCCCCCC', [1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "def fuse_on_atom_mapping(fragment_string):\n",
    "    '''\n",
    "    fuse_on_atom_mapping - Merges a series of molecular fragments \n",
    "    into a single compound by atom mapping\n",
    "    \n",
    "    ie `R1-[*:1].R2-[*:1] >> R1-R2`\n",
    "    \n",
    "    Fragments with paired atom mappings will be fused on the atoms connected to the mapped dummies.\n",
    "    Mappings that occur once or more than 2 times are ignored\n",
    "    \n",
    "    Inputs: \n",
    "    \n",
    "    - `fragment_string [str, Chem.Mol]`: input molecule fragments\n",
    "    \n",
    "    Outputs: \n",
    "    \n",
    "    - `new_smile str`: fused molecule\n",
    "    '''\n",
    "    \n",
    "    mol = to_mol(fragment_string)\n",
    "    \n",
    "    if mol is not None:\n",
    "        fragment_string = to_smile(mol)\n",
    "\n",
    "        fragments = fragment_string.split('.')\n",
    "        fragment_check = [i.GetNumAtoms()>1 if i is not None else False for i in \n",
    "                      [to_mol(k) for k in fragments]]\n",
    "        check = all(fragment_check)\n",
    "    else:\n",
    "        check = False\n",
    "    \n",
    "    if check:\n",
    "\n",
    "        new_mol = Chem.RWMol(mol)\n",
    "\n",
    "        maps = defaultdict(list)\n",
    "        for atom in new_mol.GetAtoms():\n",
    "            atom_map = atom.GetAtomMapNum()\n",
    "            if atom_map and atom.GetAtomicNum()==0:\n",
    "                maps[atom_map].append(atom)\n",
    "\n",
    "        remove_idxs = []\n",
    "        missing = []\n",
    "\n",
    "        for map_num in maps.keys():\n",
    "            atoms = maps[map_num]\n",
    "            if len(atoms) == 2:\n",
    "                atom1, atom2 = atoms\n",
    "\n",
    "                neighbor1_idx = atom1.GetNeighbors()[0].GetIdx()\n",
    "                neighbor2_idx = atom2.GetNeighbors()[0].GetIdx()\n",
    "                atom1_idx = atom1.GetIdx()\n",
    "                atom2_idx = atom2.GetIdx()\n",
    "                bond_order = atom2.GetBonds()[0].GetBondType()\n",
    "\n",
    "                new_mol.AddBond(neighbor1_idx,\n",
    "                             neighbor2_idx,\n",
    "                             order=bond_order)\n",
    "\n",
    "                remove_idxs += [atom1_idx, atom2_idx]\n",
    "\n",
    "            elif len(atoms) == 1:\n",
    "                missing.append(map_num)\n",
    "\n",
    "        for idx in sorted(remove_idxs)[::-1]:\n",
    "            new_mol.RemoveAtom(idx)\n",
    "\n",
    "        new_smile = Chem.MolToSmiles(new_mol.GetMol())\n",
    "        \n",
    "        \n",
    "    else:\n",
    "        new_smile = fragment_string\n",
    "\n",
    "    return new_smile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert fuse_on_atom_mapping('[*:1]CC.[*:1]CC') == 'CCCC'\n",
    "assert fuse_on_atom_mapping(to_mol('[*:1]CC.[*:1]CC')) == 'CCCC'\n",
    "assert fuse_on_atom_mapping('[*:1]CC.[*:2]CC') == 'CC[*:1].CC[*:2]'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "def fuse_on_link(fragment_string, links):\n",
    "    '''\n",
    "    fuse_on_link -  Merges a series of molecular fragments \n",
    "    into a single compound by `links`\n",
    "    \n",
    "    ie `fuse_on_link('R1-[Rb].R2-[Rb]', ['[Rb]']) >> 'R1-R2'`\n",
    "    \n",
    "    Fragments matching a given link are defined by substring searching. \n",
    "    Links that occur once or more than 2 times are ignored\n",
    "    \n",
    "    Note: inputs with RDKit atom mapping (ie `[*:1]CC`) will fail\n",
    "    \n",
    "    Inputs: \n",
    "    \n",
    "    - `fragment_string [str, Chem.Mol]`: input molecule\n",
    "    \n",
    "    - `links list`: list of defines linkages\n",
    "    \n",
    "    Outputs: \n",
    "    \n",
    "    - `str`: fused molecule\n",
    "    \n",
    "    '''\n",
    "    fragment_string = to_smile(fragment_string)\n",
    "    fragments = fragment_string.split('.')\n",
    "    \n",
    "    for link in links:\n",
    "        to_link = [i for i in fragments if link in i]\n",
    "        \n",
    "        if len(to_link)==2:\n",
    "            reaction = AllChem.ReactionFromSmarts(f'[*:1]-{link}.{link}-[*:2]>>[*:1]-[*:2]')\n",
    "            new_smiles = Chem.MolToSmiles(reaction.RunReactants([Chem.MolFromSmiles(i) for i in to_link])[0][0])\n",
    "            fragments = [new_smiles] + [i for i in fragments if not i in to_link]\n",
    "            \n",
    "    fused = to_mol('.'.join(fragments))\n",
    "    return to_smile(fused)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert fuse_on_link('[Rb]CC.[Rb]CC', ['[Rb]']) == 'CCCC'\n",
    "assert fuse_on_link('[Rb]CC.[Rb]CC', ['[Pb]']) == 'CC[Rb].CC[Rb]'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAcIAAACWCAIAAADCEh9HAAAABmJLR0QA/wD/AP+gvaeTAAAgAElEQVR4nO3de1yMafsA8Gum6aQDsqicSwdtFEnpgBdhKceKRWLRD5+c1i6WpRxax169ltZprejdTWxRCcUnqRxKTmVtmxTppJpqpmaa5nD//nj2nW0jm+aZmWa6vp/9I3O4rqv3tdfez3Pfz30zCCGAEEKovZjKLgAhhFQbtlGEEJIJtlGEEJIJtlGEEJIJtlH1RwgRi8XKrgIhtYVtVJ01NjaePXvWzs7um2++mTVrFpvNVnZFCKkhBi54UksvXrwIDw8/ffp0XV0dAOjr69fX1w8YMCAqKsrZ2VnZ1SGkVnA0qlYIITdu3PD19bW2tj506FBdXZ2Dg0NERMTTp09dXFxevXrl7u4eHBwskUiUXSlC6gNHo2qCy+VeuHBh3759f/zxBwDo6OjMnz8/MDBw+PDh1AdEItHu3bt37txJCPHy8jpz5oyRkZFSS0ZITWAbVXmFhYU///xzZGSkmZlZYmKiiYlJQEBAYGDgJ5988u6H4+LilixZwmaz+/Xrd/78+dGjR8ujpKysrD59+piamsojOEIdDV7UqyqJRJKcnDxv3jw3N7ejR4/W1dUZGBjExMQUFxcHBwe/t4cCwPTp0x89ejR69Oji4uIxY8bs27ePxv+ONjU1XbhwwcPDY9SoUYcPH6YrLEIdHUEd0k8//SSRSKify8vLg4KC1qxZc+vWLUIIh8M5d+6cu7u7iYmJiYnJwIEDV69e/dtvv7U9uFAo3LRpE4PBAIDp06ez2WwZqy0pKdm+fbuxsTH1l8rIyCgkJETGmAipCmyjHVFERMSyZcv2799fX19PCNm7d++DBw8KCwvd3d1//fVXMzMzqoE6Ozv/8MMPdXV17cty6dKl7t27A0D//v3v3r3bviAPHjzw8/PT1NSkGqi1tXVYWBhVNkKdBLbRjujp06dHjx79/vvva2trk5KSqBdramomTJjw4sWL/v37T58+PS4uTiQSyZjo1atX1PonFou1d+9e6fj3HzU2NkZHR0tvrTKZTE9Pz+Tk5LZHQEhtYBvtiMRisUgkampqYrPZ8fHxhJDY2NjPPvssODiYEFJRUUFjrqamJukF/owZM/7xAr+0tDQoKKhnz55UA+3Vq9emTZuKiopoLAkh1YJtVJUsWrTo4cOH8ogcGxtLXeAPGDCgtQt86vqdxWJRDXTEiBHHjx/n8XjyqAchFYIz9R0ah8NJTU398ccf37x5U19fX1VVRTU72s2cOfPRo0dOTk6vXr0aN27cf/7zH+lb0idKR44cee7cOSaT6ePjk5ycnJ2dHRAQoKurK496EFIhuG60Q6uurr5165atre3Jkyc5HM6cOXMmT54sv3QCgWDjxo3UWqWZM2cGBQVFRUWdOnWquroaAHr37r148eLAwMC+ffvKrwaEVA62UdRSdHT08uXLORwOi8USiUQA4OLiEhgY6O3tLZ2RRwhJYRtF7/HixYtt27ZpampqaGgEBgY6ODgouyKEOi5sowghJBOcYkIIIZlgG0UIIZlgG0UIIZlgG0UIIZlgG0UIIZlgG0UIIZlgG0UIIZlgG0UIIZlgG0UIIZlgG1UHRUWQkfHnz7//Dq9eKbUahDoZbKPqICUFpkyBggIAgMuXITVV2QUh1JmwlF0AooevL6xZA1euAABIJJCb2+onNTQqxeKK1t4dMmSIhoaGHApESG1hG1UTw4aBtjacPw8A0NAAQ4e2+smxY8NTU4Nbe3f//v0LFizAI+YRajtso+ojJATGjwcvLzA0hE8/bfVjJia9Pm3l7fLy8o0bN+bm5kZERMirSoTUDm6Upw5++gk4HFi7Fs6cgU2b4MABWLSoPXFevnz56aefCgSCu3fvOjk50V0mQuoJp5jUgbY26OgAAPj7/3l13z5mZmbr168nhKxbtw7/+4pQG+FoVOUlJ8P+/RAW9p4L+bw8CAho+aKFRUR+/ukWLx4/ftza2hoA6uvrraysSktLz5496+fnJ6+iEVIjeG9UtYlEsH49PHsG16+/p41yOHD7dssXGYyi2++8yuFwqB/09fW/++67xYsXf/PNN7NmzdLX15dL3QipERyNqrZDh+DLL2HwYMjNfc+1PJcLDx+2fFFT85VQWNTixREjRhgYGFA/E0KcnZ0zMzO//fbbXbt2yaNshNQJtlEVxmaDhQWw2ZCQANOm0Rn53r17Li4u2trav/3226BBg+gMjZDawSkmFbZlC7DZMHEizT0UAJydnRcsWNDY2Lhx40aaQyOkdnA0qqqePQN7ewCAx48/tEq03UpKSqytrevr65OSkjw8POhPgJC6wNGoqgoMBJEIVq+WSw8FgD59+mzatAkA1q9fLxKJ5JIDIbWAo1GVFBdXO2NGt08+gT/+gO7d5ZWlsbFxyJAhRUVFx44d+7//+z95pUFIxWEbVT2NjY02NjY9eixcseLrpUsN5JrrwoULvr6+RkZG+fn5RkZGcs2FkIrCi3rVc/DgwcLCQj4/xt9fV965fHx8xo0bx2azd+/eLe9cCKkoHI2qGMXP/Dx58sTBwYHBYDx+/Li1PU0Q6sxwNKpiNm/eXF9f7+3trbDZczs7u2XLlolEonXr1ikmI0KqBUejqoRaFa+lpZWbmzt48GCF5a2srLS0tKytrb1y5crUqVNlCSUUCquqqkxMTKg/EkKKioq6d+/erVs3OipFSAlwNKoypBsvffXVV4rsoQDQs2fPbdu2AcDatWsFAkG747DZ7PT09DNnzty5cwcAqGF1eHi4v79/bGwsbeUipFg4GlUZERERixcvNjY2zsvLMzQ0VHB2kUhkb2//7Nmzffv2BQYGtvYxJpP5gb9RTCbz5MmTDx8+DAkJMTExaWxsLCwspNZUbdiw4ddff5VP7QjJF+7wpDLS09MBYM+ePYrvoQDAYrH2798/bdq0HTt2UMvy38vJyen169etvXv9+nUzMzMXFxcGgxEcHOzl5eXg4AAAZWVleGwJUl3YRlVGfX09APB4POqPIpHoxIkTPB7vq6++klPGEydO1NbWrlu3TktLCwAaGhoAgMFg6Oq2utBKW1tbh9pB+n2YTKb01mpwcDD1Q01Nza5duw4fPkxj5QgpFEEqgrrmNTIyqqqqIoRkZmZSHa2wsFAe6SorK7t37w4AV69eJYTweLyBAwcCwIkTJ2jMUlhY6O3tff/+fRpjIqRgOMWkMmbPnu3h4cFms6k9QB0dHefPn8/n8+W0CdO2bdtqamomTpw4ZcoUADhw4EBRUZG9vf0XX3xBS/zg4OCYmJgpU6b06NEjOTk5LCyMlrAIKR5OMamSZ8+e2dvbA8CjR49sbW1LSkqsrKwaGhpSUlLGjRungES3bt0aO3YsXVn4fP7bt2+pnzU0NPr27UtXZIQUStnDYfRxVq5cCQATJkyg/rhz504AsLOzE4lENGah1vZT66sIIfPnzwcAX19fGlMgpDawjaqY6urqHj16AEB8fDwhhM/nU7csjx07RleKixcvQrObsHfu3JHrTViEVB22UdVD3UY0NzdvbGwkhERHR1Ndr7q6WvbgjY2N1Nr+8PBwQohYLB41ahQAbNu2TfbgCKklbKOqRygU2traAsCBAweoV6gbo9QR8zKidnL69NNPhUIhIeT06dMA0KdPn/r6etmDI6SWcIpJJd24ccPDw8PAwCAvL8/ExISuTZjKy8utrKw4HA61fRSXy7WysiorK4uMjFywYAGN9XcotbW1OTk5TCZz6NChhoaGbDZ769atXC5XT0/vwIEDSnnYAakWXPCkkiZOnOjp6cnlcrdv3w4AdnZ2S5cuFYlE69evlyXsxo0bORzOnDlzqCmmkJCQsrKy0aNHU1NM6qqsrOzixYsXL14sLS0FgMLCwuXLl0dGRlpYWMTExCi7OqQKlD0cRu304sULbW1tJpOZlZVFCHn79i21SdKVK1faF/DBgwdMJlNLSys/P58QUlBQoKOjw2Aw1H5tvEAgOHbs2JEjR4RC4fbt24VCoUAgePLkyaJFi548eaLs6pAKwNGoqjI3Nw8MDJRIJGvXriWEyLgJEyFk7dq1EolEun3Uhg0bGhsb/f39qSkmNaalpTVo0KCcnJySkpI5c+ZoaGjU1NRcvXpVIBB06dJF2dUhVaDsPo7aj8PhGBsbA0BUVBQhpKmpycrKCgBCQ0M/NlRERAQA9O7du66ujhBy8+ZNANDX1y8pKaG/7o5n1apVJiYm1OIEqcuXLwcHByurJKRCcDSqwgwMDKjl91999VVDQ4Ompua///1vAIiPj//YUNRX9u3bZ2hoKBaLqY3ut27d2kk2XqI2TElMTNyzZ09CQkJoaGhmZualS5eoDagQ+gfK7uNIJmKxeOTIkQCwY8cO6pX4+Ph2PNEkkUji4+MlEgkh5MiRIwBgZmbG5/NpLrej4vF45ubmpqamiYmJEokkKSnp6NGj6enpyq4LqQZc8KTy7ty54+bmpqOj8/z58wEDBsgYraamxtLSsqqqKjY2dubMmbRUqBKWLl169erVkJCQJUuWKLsWpGLwol7lubi4+Pj48Pn8LVu2yB4tKCioqqpq/PjxnaqHwv+u669evarsQpDqwdGoOiguLra2tubz+ampqe7u7u2O8/z5czs7O4lE8vDhw2HDhtFYYcfH4XCGDh0qkUgeP35M7VqAUBvh7vfqoF+/fhs2bNi1a9fkyZO7du3a7jhcLlcoFK5cubKz9VAAMDQ0dHNzS0lJSU5OnjdvnrLLQaoE26ia2LRp0+DBg/39/fl8vixxTp482UEu54uLi/v16yf9o0AgEAgEcn00c+rUqSkpKYmJidhG0UfBi3q1UlZWJmOE3r17M5nKv2N+9erVp0+f2tjYeHl5AUB4eHhsbOywYcNCQ0Pll5TNZtvb2zOZzJycHAMDA/klQmoGR6NqxcTERNkl0MPY2PjYsWOjRo0ihDAYDF9fX2dn5//+979yTWpkZOTo6FhcXHz79u1p06bJNRdSJ8ofdyD0rsbGxpMnT4pEooyMjD179nzyySeKyevh4ZGVlXXmzBnFpEPqAUejqCMyNTVdtmyZWCy+cuWKm5ubwvJ6enquWrUqMTGRx+PhA/WojXA0ijoiY2Pj33///fHjx2/evFFk3r59+zo6OvJ4vKSkJEXmRSpNIzg4WNk1INQSi8XKzc3Ny8tramrKysqqqqqKjIzMyclhs9kuLi5yTV1dXX3z5k1tbe1Zs2bJNRFSGzhTjzqoy5cvr1y50tnZOSYmhs1mc7lcANDU1JT3bin5+fmWlpbdunWrqKjQ0tKSay6kHrCNog6qoaHB1tZWKBRmZ2f37t1bkamHDh2am5t7/fr1SZMmKTIvUlF4bxR1UHp6emPGjKH2W1Jw6tmzZwMAniCC2gjbKOq4lLVdCNVGL126JBaLFZwaqSJso6jjmjRpEovFysjIqK2tVWReOzs7c3PzioqKO3fuKDIvUlHYRlHH1b17d2dnZ6FQmJycrODU1FNMq1evPnv2bF5eHk4hoA/AKaZOp6amRsbFmIaGhrLvD91GZ86c2bJly5QpU06fPq2YjADA5/Pd3d0LCgqko2ADA4Nhw4a5ubm5uro6Ozv37NlTYcWgjg/baOdy7dq1W7du7du3T5Yg1KbOq1evpquqD6ioqFi4cKGOjk5UVJSenp4CMorFYl9f35iYGGNj4yVLluTl5d2/f7+kpKT5ZywsLKZOTTM37+3kBPb20GJZFCHw8CEMHw5MJgiFUFQEFhYKKBwpDbbRToTH41lbWxcXFw8cOFBfX799QQghhYWFPB7v5s2b48ePp7fC93J1db1z586FCxe8vb3lnYsQEhAQcOrUqa5du96+fVu672pZWdmDBw+ys7Ozs7PT09M5nHo9vSYulwEALBZYWoKbG7i6goMD2NiASATa2nDsGAQEQGkpLFkC16/Lu3CkVMo5AgopA3WQ/YgRI8RisSxxQkJCAMDGxkYoFNJV2wccPHgQAD7//HMF5Nq6dSsA6OrqpqWltfYZoVD45MnzY8fIkiXExoYwmQTgr3969SJhYcTOjjg5kYoKUlJCJk1SQOFImbCNdhavX7/u0qULg8G4ffu2jKEEAsHgwYMBoMXB7nJSWFjIYDAMDAzkfVJpeHg4AGhoaMTExEhffPTokZ+f35EjRx48eNDU1PTut7hckpZGwsKIjw/p3ZsAkBMniKMjSUggfn7YRjsFbKOdha+vLwDMnz+flmgXL14EACMjo6qqKloCftjw4cMBICEhQX4pYmNjNTQ0GAzGjz/+2Pz1Q4cOSS/ddHV1XV1dv/zyy0uXil+9en+cly9JRQVxdCSEkJkzSVQUtlH1h220U0hPT2cwGLq6ukVFRXTF9PDwAIC1a9fSFfADdu7cCQBLly6VU/yUlBRtbW0A2Lt3b4u3Xrx4ER4e7u/vb21tzWAwqH5qackHIMbGxNOTBAWRuDhSXf3XV5qa/myjRUXExgbbqPrDNqr+xGLxyJEjAWDHjh00hs3NzWWxWCwWKycnh8awreUCgB49esjjbuzTp0+7desGAKtWrfrwJ2tqaq5du7Zz524vL8knn/ztluiIEX99TNpGCSEhIdhG1R+2UfV34sQJAOjXr19DQwO9kVeuXAkAEyZMoDfse1lbWwPAzZs36Q1bUFBgbGwMAPPmzfvYmbf8fBIZSVavJk5OZM2av73F4RBCSG0tOXOGHDtGX7moQ8I2quY4HA7VJs6fP0978OrqaupI9/j4eNqDt7BlyxYACAwMpDHm27dvLS0tAWD8+PGNjY00RqakpxMAMnAgkUhoj406EGyjam7Dhg0A4OLiIpHPv8rUDIy5ubk82lBzWVlZANCnTx+6fhEOhzNixAgAGDlyJJfLpSVmC2IxMTEhAOThQ3mERx0FtlF1lp+fr62tzWQys7Ky5JRCKBTa2toCwIEDB+SUgiKRSIyMjHR1dS0sLAICAiIiInJzc9vdUgUCAbWXqLm5eXl5Ob2lNrdyJQEg334rvwxI+bCNqjNqf43ly5fLNQu1b4iBgUFZWZn8suzYsYPK0vzhkZ49e3p6eu7cufP69eu1tbVtDEU97gkAvXr1+uOPP+RXMyEkOZkAEBsbuSZBSoZtVG0pprtR5N2vjx8/Ti2Mj4yMTEtLCwsL8/Pze3d7FDMzMz8/v7CwsLS0tA/cZFi3bh0AGBoaPpT/xbZQSKg5/efP5Z0KKQ22UfWksGttilzvHly+fJnFYgHA999/3+KtkpKSuLi4oKCgiRMn6urqNm+pmpqaDg4Oa9asoS7/pV/ZtWsXAGhpaSUnJ9Ne6nstXkwASEiIYrIhJcA2qp4UNvMjJae5rNTUVB0dHQDYtWvXhz9JnSF65MgRPz8/S0tL6VJ5iqmp6axZs+bOnctgMJhM5oULF2gs8sMuXyYAZORIhSVEioZtVA0pch2SlHRlVVRUFF0xc3JyunfvDgArVqxo/vqNGzeaX78LBIJ3v1tXV5eWlrZ3715PT0/p9qA6OjpaWlphYWF0VdgWjY3E0JAAkJcvFZkWKQ62UWWqr6+vbv4UISHV1dUybr9ECFmxYgUATJw4UcY4H4ta59+3b19a1vkXFr6mzlL28fFp8b8JtcWUlJ6enqurK3X9XlBQ8N5oeXl5Z8+e7dWrFwDcunVL9vI+ypdfJg4b5n7o0H8UnBcpBrZRpSkoKIiNjT1x4sSjR48IIYWFhQEBAdbW1oWFhbKEVeQzmi3Q+NRpZSUZOlTg4jJv3Lhx727sJBQKc3Nzjx8/7ufnZ2Nj0+L63cTExNPTMygoKC4urqampvkXN23apLB9AJqLiooCAHd3dwXnRYqBbVRpGhoagoKC1q9fz2azCSFcLpfH4/n7+1NtlNc6Pp/f0LqJEycqpVNQaNkDhcsljo4EgDg4iOvq6v/x85WVlQkJCdu3b588eTL1dLwUi8Xy8PCQfvL+/fvUeFlODyO0hsvl6ujoMJnM0tJSReZFioFtVGnEYvG1a9dSUlJ4PN7SpUupReBUG62srDRp3YQJE6B1+vr6hoaGVGtWiqlTp2ppabX72dOmJjJlCgEgZmakfSu1SkpKoqOj16xZ4+rqqq2tPX36dOlbEomkf//+AHD//v32ldduXl5eAPDDDz8oOC9SANYH/oVEcsVkMidPnkz9fOrUqeZvMRgManr6vbS0tFos7pEihIhEIg0NjaamJhpL/SiNjY0ikUggELTju4TA8uVw7Rr07AmJiWBs3J4CTE1NfXx8fHx8AIDP57PZbOlbDAZj5syZhw8fjo2NHTVqVHuit9fs2bPj4+NjY2OpO9dIrSi7j6O/kV7Utxs16pHf1pwflpSUBACGhobtW/P/5ZcEgBgYkAcPaC/tTykpKQBgYWEhrwStqKmp0dTU1NTUbDGpiNQAttEO4Ysvvnj16tXu3btHjx69ceNGWWaHXrx4Qa2Ez8zMpLHCtmhqarKysgKA0NDQdnx9714CQDQ1ybVrtJf2F5FIRM3XK34KjrptHRERoeC8SN6wjXYUYrG46H/q6/95XuUDvv76awAYPXq0gidSQkNDAWDw4MHtWPP/5g3R1SVMJqFv1Wmrli9fDgDBwcHyTtTQ0ND8/wLqrKcZM2bIOy9SMGyjaojD4ZiYmADAzz//rLCkb9++pWbJr1y50savTJtGkpL+/NnXl9y+raAdjhMTEwHAzs5OrlkEAoGHh8fChQulB+EVFBQwGAx9fX3pKteXuCJfLWAbVU/UnFXfvn1lHNi2XUBAAAA0X130j4YMIaNHEx6PEEJsbeVV2Luampqoh6Py8vLklEIsFnt7ewOAqanpmzdvCCFCodDT0xMANDQ0mk9O9OvXz9vb++DBg2lpabQfT4AUA9uoehKLxY6OjgAQFBSkgHSPHz/W0NBgsVjNNwH5R/b25NixP/fiVGQbJYQsWLAAAPbv3y+n+GvWrKGm2qhnKyQSydKlSwGgR48eGRkZCQkJ27ZtmzRpUotVrrq6ulOnTt28efP58+fz8/Pfe0/m+vXr0tebmpqio6OPHj36+++/y+kXQW2BbVRt3blzh/bTQFszbtw4AFi/fv1HfcvenohExMWF/P67otvor7/+CgDOzs7yCB4cHAwAOjo6qamp1CubN28GgC5dumRkZLT4cPNVrg4ODs0XCFtYWEyfPv3bb7+Ni4urrKwkhJw7d27FihWhoaF1dXWEkKCgoMOHD6emprq7u799+1YevwtqC2yj6mzevHkAMG/ePLlmiY6OBgAjI6OPXcpjb08IIZmZxMtL0W20oaFBT0+PwWC8fv2a3sjSrVEvXrxIvXL06FEA0NTUTExM/PB3eTzevXv3wsPDly9fPmLEiBaPXSxcuLCwsHD37t1nzpzh8/nPnj2TDkvXrFlz9+5den8R1HbYRtVZcXGxnp4eAEiHRbTj8/kDBw4EgGMfPz1EtVFCyIoVZPBgmgv7R7Nnz4b37WEqi3e3Rv3ll1+YTCaDwTh9+vTHRisvL09KSjp48OCiRYusra03bNjA5XLr6+vLy8vfvn179OhR6mPUE8DU+BQpBbZRNRcUFAQAw4cPF4lE8oi/c+dOata7HfGljaWykuzZQ3Nh/ygyMhIA/vWvf9EVULo16s6dO6lXbt68qa2tDXRsni0Sid7bKJuamvz9/ePi4mSMj2SBbVTN8Xg86rCNU6dO0R78zZs31Gg3JSWl3UFeviQmJmTQIPrKapva2lotLS0NDQ1a7iq+uzVqVlYWdXIUvYdCUxobG/Pz80tLS+fOnXv16lXa46OPgm1U/f3yyy8A0KtXr7Yf+tZGCxcuBABvb29ZgojFxNT0r1OIKyvJggXExYWeCj/ss88+o+U/MK9fv+7Xrx+1tJ4alb948aJ3794AMH/+fNk3kH1XeXl5aGjounXrfHx8AgICAgIC0tLSaM+C2gjbaKcwZswYAPj6669pjHn37l1qCxXZ15CvWvXXKcSKPAOO2md62rRpsgSprKy0trYGAOnWqBUVFRYWFgAwYcKE9+7Mj9QMttFO4eHDh0wmU0tLi64F5xKJhNoh6Vs6jmBvcQqxws6Aq6ysZLFY2tra7R6nNzQ0jB49GgCGDh1KbRFdV1c3fPhwAHB0dORyubTWizoobKOdBbX828vLi5ZoP/30EwD06dOHlk7RYgQaF0ft2Sx74PfYsGHDwYMHpX8cO3YstQ6/fZfed+/e1dPTMzMzo3a0oh4ApTYWqKiooK1o1LFhG+0sKioqunbtCgCyz0hwuVzqlKRz587RUhv5+whUfmfA7d27FwC0tbWlRzZNnTqVuolpYGBAPeoeHR1NbaHdRvfv38/PzyeEiMViapNTU1NTGXc7RKoF22gnsn//fgAYMmSIdLOM9vnmm2+oR4Bo3EGqxQh07lwCQP79b7rCE0LIuXPnqNOVpTvzU1suaWhoUDu5NGdmZvb555+HhYXdvXu3jRtWUQ+Adu3alXoAFHUeDEJI63s6I7XS1NRka2v78uVLQ0NDLS2t9gUhhDQ0NPB4vLt37zo5OdFVm0AAvXoBhwMvX8KgQXDpUs2hQ9cNDW/Gx5+kJX5CQsKsWbNEIlFYWNjatWsB4PLly3PmzJFIJCdPnqQOccnKysrOzs7Ozs7IyKipqZF+l8ViWVpaurm5Uc9rvnuIHgAEBwfv2LFDR0cnKSnJ3d2dlpqRylB2H0cKlZ6evmPHDhn/znh4eMhjC74FC0SOjuUnTmQQQrhcrq6uLl1nwN27V6WvbwjNJsRSUlKopfIh75vJEolEubm5ERER1KPumpqazX99Q0NDV1fXTZs2SR91f/cBUNSp4Gi00+Hz+XV1dbJE0NbWppaa0ys6Onru3Llubm5paWkAMGPGjLi4uB9++EHGw4t++w3c3cHK6rWdXWh4eBiDwcjJyRkzZkxtbe3KlSup6/oP43K5Dx48uHfv3v379zMzM8vKyqRvMRiMAQMGUA/mHz9+nNoQGnU6yu7jCP2pxQiUWgwwadIkWWK+ebb+rFkAAAMTSURBVEP69ycAxMuLCIWEEPLy5UvqTujcuXPbNztfUlISFxcXFBQ0ceLELl26AMC8efN27dolS51IpWEbRR3I9OnT4X+nEMt+BlxVFbGxIQDE2ZlQGyJXVBAnp+lA38J4oVCYnZ1NrbpHnRZTqUNhhP5m1qxZABAbGwsA3bp1Gzt2rFAoTEhIaEcoPh9mzoTffgNbW0hMhC5dgMuFqVOhrOziZ5+tio2NbfckW3MsFmvEiBEfOA0bdQZ4bxR1ILW1tdSxneXl5UZGRrGxsXl5eZ9//jm1u8pHyc2FceNAXx8yMqBPHxAKwcsLrl8Hc3NITwdjYzlUjzorbKOoY/Hw8Lhx40ZERMSiRYtkDPX8OTCZYGUFhMDixXD2LPTsCenpYGlJS6UI/Qkv6lHHQu2mHBMTI3uoIUPAygoAYMMGOHsWDA3h2jXsoYh+OBpFHUtFRYWpqamGhkZ4ePjYsWOprZJk8d13sHUraGlBQgJ4eNBSI0J/g20UdTi7d++OjIzMy8sDAENDw6FDh1JPEDk7O/fs2fOjQp07B/7+wGBAVBT4+MinXNTpYRtFHdGhQ4du376dmZlZWloqfZHBYFhaWn722ZrBg1c5OYGdHfz98SIAAD4fGAygZs55PNi8Gb7/Ho4ehVWrFFg96mSwjaIOrbS0VPqc+507d3g83tixl1JTZwCApiYMGwauruDgAA4OYGMDDAZs2QJXrkB2NrBYsGwZLFsGAgGMHavsXwOpNWyjSGWIRKKcnJynT3ulpPS5fx/y8qD5X15jY3ByAgsLePwYPD1h7do/26izs/IqRp0DS9kFINRWLBZr+PDhw4eDvz8AAJcLT55AdjZkZEBqKpSXQ2Ym2NhAYCDs3g3e3souF3UaOBpFaqKgAN68gevXwd0dWCw4dQoMDHA0ihQB140iNWFu/tc9UA8PYDDg3j2lFoQ6DWyjSD2FhcGbN8ouAnUOeFGP1EpJCejrQ9euAAAFBWBsDHp6yq4JqTtsowghJBO8qEcIIZlgG0UIIZlgG0UIIZlgG0UIIZlgG0UIIZn8P/p4OX8YijIWAAAB+npUWHRyZGtpdFBLTCByZGtpdCAyMDIwLjA5LjEAAHice79v7T0GIOABYkYGCJCH4gZGNoYEIM3EhJ1mZmTQ0QBpYJRg1AFp1sKlHq6OGZs6RmY2hwyQecyMcAa7gwaYgSQDFmCBG8UENQqPVjQdMEdCxTHkoY7DqR/TRogCmAZuYAgyMjEwMYM0MbMysLIxsHMwsHMycHIxcPNkMPHwJvDyZTDx8SfwCzDwC2YwCQlnMAmLJIiIMoiIZTCJiSeISzCIS2YwSUolSEkzSMlkMMnIJsjKZQDDiUGQO0FOKEFWLEGEhY2BjZWZiZGNW5Cfj5eHTUhOVkxEmE1cUkpGVkwcFokM8nm3v9mregotAXHybk+zf+i2bD+I7TlpyT4Ye5JK5/5Gc31ViLjKge250apIauyR9ILZDwok97f033QAsb8l/92vdOSYA1TcXnPrHpi4fXfDeQeo3v2FysJQ9rf9D93S4OKaW+c4QO2yg7GB7rGHqQG6xwHBXmIHM6f1TITDBv7vYPdvz7V2YOpkALMtnlY7iDHIgN2Z1RbnMDeVCaz+yN01DgueTrUD65U86/Bl72awGrmsfQ6TXEzBatay/HEINO8Gi3NFsTpW+a/bB2LPOy/jyGyqATY/uOKVQ9aNnWC2du9uh+OBE8Hs7MDZDlqn+Q+A2GIApOuV1tCXdQcAAAHYelRYdE1PTCByZGtpdCAyMDIwLjA5LjEAAHicpVRLcuUgDNz7FFwglD4I0HrerFLJIovcIfvcv0Y2WHbmJZVEdrHotnGjbhBLWp+X2+Pbe/KHbsuSGNeR4NOhqumVAGBZ52NW6tXAA2R7B+u7if6kryTOY6pgGyqYSTWmYhUwYbmqghnappIuOHqgLF+pvNyr4D3+6ChdcHSke0XFahHVrZaam/YeTVeb1qFC+CGhb3Lhk6OjFsldG0ZzOWopvZSooz1dunTqZh8lNm895uhUS/k/l+ef53KcujuVX6W799EFR+noozuVb84Lferowh5ZBdybzN1ioKij48YUoOAesVUgOv6toLGzayqlNRhVAW/efn9LWQ8W90FmLlaL9eD0gYZKMF3JKiDDUWfi2P3Ssm30vDGhtxrbo56Zi4yd5gISc6S5A7ShUqUFO6BZBzQaZ7cIajTdXhHmTcPCMRW7VYQ3R9aNpfZIurjZw0Fo271JDBUnhuRMqhNDzYmVoCeCsBNDiE4oIY9YlxWhr2MIxb9IQl/HELYz6T5NE8FODJGvY4joTHxRQ+SLGiI5k+rTaiJf1NDakgdRn6aJ3akhxv3LjGZW3Vfn+zRcCz/+oT34p5T+Pt+Wf/38mJuPEr0YAAABpXpUWHRTTUlMRVMgcmRraXQgMjAyMC4wOS4xAAB4nG2RvW7jQAyEXyVlHKwJ/u1yKVcHtbnD9QcXjuIusoMgKQ7Iw4crW4LjuBNnvyGHVE9939//e+hou+p7gqiikC0MNByGob3wdjVQ1IfhzA08fcl2dViAgQ9093lP4FyzpbUAGaPt12RpQ1FMKgG7o6bNGkGYlP1SI0AjlUSQya2Et4TKkM+qo7Ce1cW/2OchF4KbF0oFtKqGgJDdxTxlqG5EU5uzVMC8mk0pZhsT6SnXqTNfJD2tKWGvWC8YnVsvARdlXm8xzZstwuJZJglINa3tSi7YbsnzhSM6Mjo1yCVXSwgFY5nSFLXiOfqgmOa0yaABN4SdrcSsuEqD41yKytIQz6jNU4Ul2hpUy9NBsVoJooKItjCRUxu6caiIxZtSsmH25hI2mv6rxm/k1rhGrmkHktw6x0myNFvsqcVqWqXd+3H8+3Z87RSeP8bx/+Puaf8CD1GOx5df8fh79/rnY3zavwF1hN+hqH9Q0lG5ospPijum71TUNybyVSy+kUs6tivKbkz8/AI6oNZrY2R+rAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<rdkit.Chem.rdchem.Mol at 0x7fe63ca1fbc0>"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fragment_smile = 'C1CCC([*:1])CC1.C([*:3])CC.c1cncc([*:2])c1.c1nc([*:1])c2c([*:3])nc([*:2])cc2n1'\n",
    "mol = to_mol(fragment_smile)\n",
    "mol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAcIAAACWCAIAAADCEh9HAAAABmJLR0QA/wD/AP+gvaeTAAAgAElEQVR4nO3dd1xTV/8H8G8IG1Q2PlLBqgWLA2UpoowYwUGL1uJqrbQ+UrV1tLWF36MtdEips3WCWysiOB+KMhJEFGtVUByVKkOGCElBUYYBEu7vj0NjHkAc5N6bkO/75R/mGu75Rl58uPecc8/hUBQFCCGEXpUW2wUghJB6wxhFCKEuwRhFCKEuwRhFCKEuwRhFCKEuwRhF6JkSExO//fZbtqtAqk6b7QIQUlHHjx+fOXNmU1OTh4eHn58f2+Ug1YVXowh14MiRIzNmzGhqalq+fDlmKOocxihCbSUkJMyaNau5ufmrr75as2YN2+UgVYcxitD/iI+Pf++996RSaWho6E8//cR2OUgNYIwi9FRcXNz7778vlUojIiKioqLYLgepB4xRhFrt3r2bZOh3330XHh7OdjlIbWCMIgQAsHPnzvnz57e0tPzwww9ff/012+UgdYITnhCC7du3L1y4kKKoDRs2LFu2jO1ykJrBq1Gk6aKjoxcsWIAZil4ZxijSaBs2bFi0aBEA/PLLL0uXLmW7HKSWMEaR5lq3bt3nn38OAJs2bVq8eDHb5SB1hTGKNNSaNWuWL1/O4XC2bNnyySefsF0OUmM4xIQ00U8//RQWFsblcnfu3BkcHMx2OUi9YYwijSPP0N27d3/wwQdsl4PUHt7UI83yzTffkAzdu3cvZihSCrwaRRpk5cqVq1at4nK5+/fvnz17NtvloG4CYxRpiv/85z8//vijjo7OoUOH3nnnHbbLQd0HB/epR5pAJBINHz68urr6yJEjb7/9NtvloG4F+0aRRjh69KhIJJo4cSJmKFI6jFGkEYYNG0ZRVGFhIduFoG4Ib+qRRpBKpWZmZnV1deXl5f/617/YLgd1K3g1ijSCtra2p6cnRVEZGRls14K6G4xRpCl4PB4AYIwipcMYRZqCxKhAIGC7ENTdYN8o0hQtLS3W1tZVVVVFRUWvv/462+Wg7gOvRpGm0NLS8vLyAoDTp0+zXQvqVjBGkQYh9/UYo0i58KYeaZC8vDxHR0dra+uKigoOh8N2OaibwKtRpEHefPNNGxsbkUiUl5fHdi2o+8AYRZrFx8cHANLT09kuBHUfGKNIs6ja7FGxWCyVStmuAnUJ9o0izVJaWmpnZ2diYlJVVcXlclmspLKyMjg4uKmp6dKlSx4eHgEBAYGBgf369WOxJPRqMEZVFEVREolEKBT6+Pj06NGD7XK6lQEDBhQVFV2+fNnV1ZWtGsrKyng8XkFBgYmJSU1Njfy4g4ODv7+/v7+/t7e3kZERW+Whl4I39aqIoqglS5b4+vq+/fbbpqamrq6uYWFhWVlZLS0tbJfWHYwbNw5YnfZUWlrq6+tbUFDg7OxcUFBQUVGxb9++oKAgU1PT27dvb9y4cfLkySYmJq6urhERETk5OXito+oopGJkMtm///1vANDT0xs2bJi29tMdCiwtLWfPnr13796Kigq2y1RjcXFxAODv789K68XFxf379wcAFxeX6upqxX+SSqXZ2dlRUVF8Pl9HR0fx+x4UFBQTE3Pv3j1WakadwxhVLTKZ7MMPPwQAQ0NDgUBAUVRdXZ1AIFiyZEmbXrP+/fuHhIQkJiY+efKE7arVjEgk4nA4hoaGEomE4abv3r1LnkMdPXr0o0ePOnlnbW1tYmJiSEiIra2t4vfd0dExNDRUIBAwXzx6FoxRFSKVSufOnUsyND09vf0bCgsLY2JigoKCevbsKf+5MjAw4PP5UVFR2dnZLS0tzJetjoYMGQIAZ8+eZbLRO3fuvPbaawDg6en5+PHjF/9C+fddsZfc0NBQ/n2nr2b0IjBGVYVUKp0zZw4AGBkZnT59uvM3Nzc3Z2dnh4eHu7i4KD6N07t37zlz5iQkJDx48ICZstXUkiVLACAiIoKxFv/66y8bGxsAGDt27EtlqKKGhobU1NQvvviC/BpQvDVZuHDhnTt3lFszekEYoypBKpW+9957AGBsbHzmzJmX+lqxWJyQkBASEkJ+Sgkul+vi4kLu/pqammgqW32dOHECALy8vJhpLi8vr0+fPqTF2tpapZxTJBKR77t8Mf9ly5YVFxcr5eTopWCMsq+pqYns99urV68LFy7Ij8fHxz/3slSRTCbLycmJjIz08fFRHKAwMzObN28eDYWrsZqaGi6Xq6urW1dXR3dbt27dIknn7e1NR3MymezixYtOTk4AsG3bNqWfHz0XxijLGhsbp06dCgAmJiZ//PGH/HhsbCyXyzUyMrp79+4rnLa+vl4gEISGhrq4uACAg4PDokWLlFZ0t+Dm5gYAKSkptLaSm5traWlJJgY0NDTQ11B0dDQATJ8+nb4m0LNgjLKpsbExMDCQZOilS5fkxw8ePEjmOYWGhna9ldzcXC0tLWauvNRIWFiYsv6Hn+Xq1asWFhYAMHHiRLonVOTn5wOAubm5TCajtSHUHsYoayQSCdkz3dTU9PLly/Lju3bt0tLSAoDvvvtOWW2RK6/U1FRlnbAb2LVrF4fDmT17Nk3nz8nJMTc3B4DJkyczMznJzs4OAHJzcxloCynCGGWHRCIJCAgAACsrq+vXr8uP79ixg2ToDz/8oMTmGLjyUi83b960srLicDgnTpyg4/zZ2dlmZmYAEBAQwNgEz+DgYABYv349M80hOYxRFtTX1/P5fJKhN27ckB/fvn07ydDIyEjltpiamgoAbm5uyj2tmlIc81HWuLmiy5cvm5qaAsC7777L5DSJ/fv3k+BmrEVEYIwyrb6+nqzVZm1tffPmTfnx6OhoDofD4XB+/vnnrrciFosjIyNXrFghb1RPT4/L5eJ8Unl/JU1jPllZWeThiOnTpzc3Nyv9/J0oLy8HgB49euAUN4ZhjDKqrq7O19eXzJP/888/5cfXr19PMnTjxo1KaaiyspLD4RgbG8t/ory9vQHg+PHjSjm/mrpy5Qrpr6RpzOfcuXPkQaOZM2cynKGEg4MDAChOm0MMwBhlTm1tLVl6vW/fvvn5+fLja9euBQAOh7N582YlNufo6AgAWVlZ5GVERAQALF68WIlNqBf5mM+kSZPoyNDMzExjY2MAmDVrFisZSlHUokWLlN6xjp4LY5QhNTU1o0aNAgBbW9uCggL58dWrV5MM3bp1q3Jb/PTTTxWH+8+dOwcAgwcPVm4r6oLuMZ8zZ86QDP3oo49YnHJ0+PBhAODxeGwVoJkwRpnw8OHDkSNHAoCdnV1hYaH8eFRUFHlwc8+ePUpv9NixYwDg4+NDXjY1NRkbG3M4HA1cZO/8+fOkv5K+MZ9ly5YBwMcff8zutM2qqiotLS19fX1ap/qjNjBGaVddXe3s7AwAAwcOLC0tlR+XZ+i+ffvoaPfhw4dtnnecMGECABw8eJCO5l5cWRm1ejUlv6v+73+p27dpbE7eX0nfmM+jR490dHR0dHTq6+vpOP9LGTFiBAB0uEIYogmufk+7R48ePXnypG/fvhkZGX379iUHv/7667CwMG1t7QMHDnzwwQd0tGtiYjJ8+PCmpqbff/+dHCEzBFhc9Z0oLoawMIiMbH156BD8+SddbZ07d27SpEm1tbUzZ86MjY1VXANbiXr27DlkyJDm5uY//viDjvO/FBX5LmsUjFHaHTlyJC8vb+zYsWStSaJ///46Ojp79+6dOXMmfU2T3TLku2Cqzg+YlxccOgS3b9PbSmZmJsnQ2bNn//rrrzRlKKE6/7eqU4kGYftyuPvLyckBADs7uzbHS0pK6G46JSUFANzd3clLmUxGpkwWFRXR3XQnzp2jJk6k4uMpX1+qpYWaNYs6dkz5raSmphoYGADAvHnzGOivPHnyJAB4eHjQ3dBz1dbW6ujoaGtrd766PlIivBql3fDhwy0sLEpKSu7evat4vM3mEHQYO3asnp5eTk4O2XtSS0vLy8sLVONSZfp04HIhNrb1ZXAwrF0LN24o5+QpKSmBgYFPnjyZP3++/NkwWnl5eeno6Fy+fPnx48d0t9U5Y2NjNzc3qVRK5mYgBmCM0o7F8DI0NHR3d5fJZGfPniVHyOR/VYhRANi8GcLDob4exGLYtw++/BKGDQNra5g+HbZvh/v3X/G0ycnJU6dOlUgkISEh0dHRDGQoABgbG7u7u0ulUvl/NYtY3/pU02CMMoHF7qo2TZMfMDKMy3wxACCVPv27gwPMmgWnTkHPnhAXB8HB0KcPiMVw+DB8/DH07Qvu7rByJZw9+z9f1bmTJ0+SDF2wYAFjGUqQ/2p5TzSLyC/L9PR0tgvRGGz3KmiEW7duAYC1tTXzW85lZmYCwNChQ+VHyF4jio+iMubMGWrAAOrwYerDD1uP1NdT7u6U4gJ+169Ta9ZQ48dT+voUQOsfe/umKVOmbN26VXHWbXuHDx8my/5//vnndH6OjpEAdXJyYr7pNiQSiYGBAYfDEYvFbNeiETBGGULCS3EtEmY0NjYaGRkpzronmz5t2rSJ4UpSUykDAwqA+r//e6H3NzRQAgEVGkq5uFBjx+bJf/GTnaUTEhJqamoU35+QkEAydPny5bR8gOdpbGw0NDRUkfAiS4glJCSwXYhGwBhlyPvvvw8Aylp55KX4+/sDQFxcHHm5a9cuAJg6dSqTNSQnt15dzp9PvcKweWnp/Z07dwYFBZEHOgkdHR0fH5/IyMicnJy4uDgyn+mrr76iofwXpTrhtWrVKgBYsGAB24VoBIxRhuzevRsApkyZwnzT5HGp+fPnk5clJSUAYGJiIpVKmSng1KnWDP34Y6qTXo2aGsrFhQoNpdLTqWc99S6VSi9cuPDtt9+OHj2ay+XKI1VfXx8AwsPD6fkELyoyMlJFwos8CGBvb892IRoBY5QhJLwGDvRkKrueunz5MgAMGDBAfqR///4AoLhzCX2Skig9PQqAWriwswylKOro0aedoYaGFJ9PRUVR2dnPfH9tba1AIFiyZImVlZWxsfGoUaOUXvzLUp3wkkqlJiYmAFBWVsZ2Ld0fxihzfHxqASiFnesYIpVKTU1NLS0tq6qqyJH58+cDwE8//UR300eOUDo6FAD1xRfPf7NEQgkE1JdfUk5OFIfzNFKnTv09JCTk6NGjbTpD5VQwvFRhv/i33noLAGhasQEpwglPzHnjDWMAYH7WE5fLzc3NFYlEZLVNYGr26OHDMGsWNDfDl1/C2rXPf7+eHvD5sHo15OZCZSUkJEBICPTpA/n5e7Zv3z5t2jRzc3NXV9ewsDChUNjc3Cz/QldXVxMTkzt37ty7d4/Gz/MCuFzu2LFjAeDMmTPsVtLS0vLgwYNBgwaR7VIQvdjOcQ0SF0cBUP7+bNdBUSKRiMPh6OnpZWZm0tRDGhd3aOjQPwColSu7dB6ZjMrOzlm1apW3tzcZiCesrKwUl15WnSuvDRs2AMDcuXNZrEEqlZIhTWNjY8VFxRBNMEaZIxJRHA5laPjM8RPGnDx5Uk9Pjzxfb2xsHBAQEBMTo8T70P3793O53F69TFavLlfWOSmKqq+vFwgEoaGhjo6Oo0ePVvwnVQgv4vr16wBgY2PDVgHyDDUyMsrIyGCrDI2CMcqoIUMoACozk80ajh07pqurCwAuLi729vaKtyaDBg1aunTpqVOnurJuZmxsLBlD//bbb5VYdhuPHz9WfHnt2jV2w0uupaXF2toaAG7TuorqM0ilUjIvuFevXr///jvzBWgmjFFGLV1KAVAsTsuRP+cjn6NeVFQUExMTFBRE9gQmtLW1PT09o6KisrOzX+rJq507d5LnL7///nt6PkHH5OF1584dJtvt0IwZMwBg27ZtDLfb1NT0zjvvkAzFXe2YhDHKqBMnKADKy4ud1uPj40mGdjhHXSqVZmdnR0VF8fl8xaU5raysgoKCYmJiysufc4cuX0tp1apV9HyCzkyfPp2V8GovJiYGAIKCgphstLGxccqUKQBgYmJy8eJFJptGGKOMqqmhuFxKV5f6Z18P5hw6dIiEY2ho6HPfXFVVlZCQEBIS0mY1P0dHx9DQUIFA0H5XuOjoaLJH9IYNG+j5BM8RHR0NANOnT2eldUX5+fkAYG5uzti+TI2NjYGBgQBgamp6ifkpdRoPY5Rp7u4UAJWSwmijBw8eJBkaERHxsl9bWFhI7vrJzpeEoaEhn8+PiooiS5xs27aNZOgvv/xCQ/kvhPnw6oSdnR0A5ObmMtCWRCIhExUsLS2vXbvGQIuoDYxRpoWFUQDUC1wRKs2uXbvIvbZ8s+VX09DQkJKS8tlnnw0ePFjxErV3794AwOFwWL+hZjK8OhccHAwA69ato7uhhoaG8ePHk76X69ev090c6hDGKNOuXKHWraPy8p75hpQU6hlP67yKHTt2kAz94YcflHZSiqqsrExISJgzZ46ZmRmHwzEzM1uxYoUSz/9qSHitX7+e7UKo/fv3A0BAQACtrdTX15PFUKytrW/cuEFrW6gTGKNMs7Cg/vOf1r/HxVFtZjreu0cBUFxu6yIdAgHVlW3V5WM+kZGRr36WTkmlUtIrpznh9SLKy8sBoEePHk1d+f51qr6+niwUbW1tzfwCjEgRxijTLCwoBweKLJrcPkZzcykvr9bn0Mkfc3NqxgwqNrb03r17L9WQfMzn559/Vlr1HWkfXjU1NYmJicxvqcZAeL0giURiamraq1cvW1vbJUuWdDgo1xV1dXXkid7evXuzsgI3UoQxyjQLCyohgfLyolpaOohRoq6udcViR8fWMPX2XgoKKxa3mXze3vr160mGMrDCafvw8vb2BoATJ07Q3XR7Dg4OAMDurMmGhoYJEyaQ54jkPchGRkaTJ0/euHFj16fl19TUeHh4AEDfvn3z8/OVUjPqCoxRpllYUBIJNXEitXt3a4wWFlLV1c98f34+tWUL9eGHSxUHyg0MDPz9/detW1fX0cyptWvXkjGfzZs30/hJFJDwkj82Ex4eDgBLly5lpnVFCxcuVHpH8Eupr6+Xj/lcvXr13LlzoaGhLi4uHA5HcVBuzpw5CQkJDx48eNnz19TUjBo1CgBsbW0LCgro+AjoZWGMMo3EaEEBZWtLRUdTc+dSU6ZQWlpPO0MbGzv+Qvn0eE9PT9LjaWRk1P5WcfXq1SRDt27dSvuH+ceiRYsUw6v9BlCMOXz4MACMGzeO+aYpiqqvryebBlpZWbUZ8xGJRGQqbp8+feR5yuVyXVxcyFTcF+mIePjwobu7OwDY2dkVFRXR9jnQy8EYZRqJUYqiIiKogQOpuXOpqVNbFzYmf0xMqGnTqJgYqpOlQsRicWxs7Nq1a9scJwvdc7ncPXv20PgZ2iHhxePxyMv2G0AxpqqqSktLS19fv6GhgeGm6+rqXnDM5+bNm+RpMT09PXmkyteIKSkp6fCrHjx44ObmBgD9+vW7e/cuLZ8BvRKMUabJY1QioeztW/tG6+qopCRq8WLKweFpngJQDg7UkiVUWlpOhzfvbcgzlPn14tqHl5+fHyhsAMWkESNGwD+bSDOmrq7Ox8fnZcd85GtWubi4KE7Fbd8J/uDBA1dXVwCwt7d/2cFGRDeMUab9979P93S7dauDxfArKqh9+6igIMrUlEx+kpqammlra7u4uISHhz9rrZBvvvmGZOivv/5K8yfoWJvwIpkeEhLCfCVffPEFAKzs4kKnL0MpYz4FBQVbtmwJDAzs0aOHYie4n59fRETEoEGDAMDBweG5Kxsg5mGMqq7mZiori4qKKh85cqTi9m1kgOLAgQMikYi8c+XKlSRDDxw4wFa1JLzkk/AvXboE/7sBFGNOnjwJAB4eHsw0V1NTM3LkSCWO+bTvBAcAU1PTwYMHV1ZWdv38SOkwRtVDdXV1fHz8vHnzXnvtNXmeamlpOTs7k+sgXV3do0ePslhhm/AiG0ABAPO9eLW1tTo6Otra2gxMXFUc8yksLFT6+cVi8cGDB4ODg69cuSL/rYlUDcao+pGvFULu/gYOHGhoaHj8+HF2q2ofXmTdtt27dzNfzOjRowEgKSmJ1lYUx3xw3FyT4ZZ26kc+/iAWi1NTUzdt2nT37l2SWSwyNjZ2c3OTSqVnz54lR5jZOK9DZMSc1qb//vtvHx+fy5cv29vbZ2Vlvf766/S1hVQcxqga09fX9/PzmzBhgpWVFdu1AACQKZMZGRmKL9PT05mvhExQ37t3748//piTk9PS0qLc84vF4nHjxl2/ft3BwSEjI8PGxka550fqhUNRFNs1oG4iIyODx+M5OTnl5uYCAEVRffr0qayszMvLIwPNzKipqfHz87ty5YpMJiNHLCwsfH19+Xz+pEmTFDuXX41IJOLz+Tdv3hw0aFB6erridHqkodjuVUDdh0QiMTAw4HA4YrGYHJk1axYAbNmyhbEa5GM+tra2W7duDQkJIYuQEhwOx8nJ6csvvxQKha+2VkhFRYWjoyMAvPnmm/fv31d6/UgdYYwiZSLLXyYkJJCXO3bsAIBp06Yx07p8jvobb7xRVlYmPy4flOvZs6c8Ug0MDMgC/i++bV9paenAgQMBYPjw4X///TdtnwOpGYxRpEyRkZEAsGDBAvKysLAQAMzMzBjY2EMsFg8bNgw6fc6nubk5Ozs7PDz8FdYKKSkpGTBgAACMGDGiqqqKts+B1A/GKFKmP/74gwSZ/Ei/fv0A4MqVK7S2KxKJhg4dCi/znI9YLCZrhSgOED1rrZCSkpL+/fsDgLOzM2YoagNjFCmTVCo1MTEBgOJ/Flb56KOPAGDNmjX0NVpZWTlkyBAAGDRoUPv+ys2bN3/22WfJycn19fXPOkPna4XI5zO5uLhUd7KmIdJUGKNIycgulXv37iUvL1y4oPjcqtIpjvl0uKAU6S0FAG1tbU9Pz847Q2traxMTEz/55JM33nhDnqdkU9XRo0czv54/UgsYo0jJNmzYAABzO1zWX9nKyspI3nUy5nPu3LkVK1a4urrKn08HgD59+gQHB8fFxXUyUlRYWLht27bAwEAjIyMul4vPYqJnwXmjSMlu3LgxbNgwGxube/fu0dpQaWnpuHHjCgoKRowYIRAIzM3NO39/dXX16dOnhUJhcnJyWVkZOailpTVixAg+n8/n8728vHR1ddt/4ciRIy9dupSSkuLv76/8j4G6AbZzHHU3LS0t1tbWAND1TYc6IR/zebX+SjIFKiAgQF9fX/6zYGRkRKZAtVkwNCwsDAC++uor5ZWPuhWMUaR8M2bMAICZM2fStOFacXExGfNxdXV9he2MFDU0NMgXTpZPgbKzs1N8T1paGmmrS0Wj7gtv6pHynT59evv27fHx8QDQv39/css8fvx4MojfRcXFxb6+vsXFxZ6enqdOnVKcUd9FZWVlaWlpqamp/fr1I1taEQ0NDWZmZlKpVCwWm5mZKas51G1gjCJaCIXCnTt3CoXC6upqckRHR8fDw8PPz8/f39/Z2VlxwOfF3blzh8fjlZeXjxkz5tSpU4oLxdPKx8cnMzPz2LFjU6dOZaZFpEYwRhGNWlparl69KhQKhULh2bNnm5qayHFzc3Mej8fn8ydOnNi3b98XPNvt27d5PN79+/fHjh176tQpxR2n6fbdd9+Fh4cvXrx448aNjDWK1AXGKGJIfX39hQsXhEJhYmJiXl6e/Di56w8ICBg/frzigE8bf/31F4/Hq6io8Pb2TkpKYjJDAeD8+fNjxoxxdHT8888/mWwXqQWMUcSCoqIicomalpb26NEjctDAwMDT05N0pDo7Oys+856Xl8fj8SorK/39/Y8fP25gYMBwwVKp1MzMrLa2try8HFfGQ21gjCI2SaXSa9eu/fbbb0lJSVevXpWvr2xtbe3l5RUQEBAQEFBaWjp+/PiqqqoJEyYcP368kytWWk2aNCk5OTk2Nnb27NmsFIBUFsYoUhUikUggEKSmpgoEApFIRA5yuVwDA4O6urq33nrr8OHDis+8Mywm5lRsbLmz8+Sff8arUfQ/MEaRKioqKiKXqGfPnrW0tDQ0NLx582aHjxgx5soVcHEBOzsoLmaxCqSKMEaRSiNjO3Z2dsUK6SWVSiUSCcOjTC0tYG0NVVVQVAS4fx1ShFvaIZXm4eFhYWFRUlJy9+5dcmTz5s2mpqabNm1iuBItLfDyAgBgY48+pNIwRpFK09LS8vb2BoUdRq2trevq6ljatxkA4J+dTxFqhTGKVF2b/e55PJ6WllZWVpZEImG4EhKj6emAPWFIEcYoUnXy/e5JP765ufnQoUMlEgnZsIRJb74JNjYgEsGtWwy3jFQaxihSdYMGDbKxsRGLxbf+SS8ejwcK16dM8vUFAGCjZaS6MEaRGiD39fLuUYxRpFIwRpEaILmZ8c/gjre3t46OzsWLFx8/fsxwJXw+AMCZMyCTMdwyUl0Yo0gNkO7RM2fOyGQyAOjRo8fs2amDB/99/rwRw5XY2oK7O4wbBzU1DLeMVBfGKFIDtra2PN4SJ6dfc3NbR+dfe8332jWT06e5DFeSkQFeXhAfD2Tnp/BwqKpiuASkcjBGkXoYOPCXzMwAobD18pP0UTI/E/7mTdi8GaKjW18mJEBtLdM1IFWDMYrUA5mzKR/b8fQEAwO4do2Fi8E5c2DtWqioYLpdpLIwRpF68PUFDgeysqCxEQBAXx88PKClBTIzma7EwgI++ww++4zpdpHKwhhF6sHKCoYMgYYGuHix9Qi5r2fl0cxPPoHbt0EoZKFppIIwRpHaaHNfP24cAG3doyIRfP/9Mx/65HJh61b44guc9oQAMEaRGmkTo25u8NZbMHcu/LNkvtJUVICvL3zzDURGPj1IUfD3309feniAmxvk5yu5aaSOcL1RpDYePQJzc+By4cEDMKJtwmhlJYwbB7dugZMTCARgaQkAQFGwbBkcPw4rVoC2NsybBwBQVQVz5sCePdC7N13FILWAMYrUyciRcOkSpKSAvz8t5y8rAx4PCgpg+HAQCMDCAgCAomDRIoiOBn19SEyE8eNpaRqpL7ypR+qE9IfS9Eh7aSn4+kJBATg7g1D4NEM//RSio0FPD7G+ePUAAAMwSURBVA4fxgxFHcAYReqExwMOB0pLlX/m4uLi+fP/LiwEd3dIT299SKmlBebNg61bwdAQkpIgIED57aJuAGMUqZPHjyE1FeLiAACqq+HECeWctrCw0Nvb+/z5oTNmiNPSwMQEAEAmg48+gj17wNAQfvutdVEShNrDGEXqZP16+Pe/oa4OAKCsDKKilHDO/Px8Hx+f0tLS4cMHbt+u36sXAIBMBh9+CPv2gZERJCW1ThJAqEMYo0jNuLlBRITSznbnzh1fX9979+6NGTMmOTm5Z8+eACCTQXAw/PorGBtDUlLrPH+EngVjFKmZ5cvhxAm4fl0Jp7p9+7avr295ebmXl1dycnKPHj0AoLm5OTj40+zshl69QCAAHx8lNIS6N4xRpGYMDGDtWliwoPURo5QU+OYbOH8epNKXO09TU9OECRPu37/P5/OTk5PJrvdNTU3Tp08/cGBLY6NXerps1CgaPgDqdjBGkfqZMgXMzCA2FgDgwAH4/nsYMwZMTGD8ePjlFygufqGT6Orq7tixY8qUKYmJiYaGhvBPhp44ccLExCQ+fpuLC9OLmSI1hdPvkToZMwa2bAEnJygqAhcXcHCAH3+EEycgNRVu3376Nnt78PcHf3/w8XnR552ampqCgoISExNNTU3T0tJcXV1p+gio+8EYRepk4kRYswaGDAEAiIyEjAwQCFr/qaQEUlMhLQ3S05/u8KGrC3PmrH3jDZm/v7+TkxOHw+nwtI2Nje+++25SUpKlpaVQKBw2bBgDnwV1GxijqLuRySA3F4RCEArh7FnK1PQ1keg+AFhaWvr4+PD5/MmTJ9vY2Mjf39DQEBgYKBQKrayshELh0KFD2asdqSWMUdSdPXwoEwqPpaampqWllZWVkYNaWlrDhw/39/f38/MbMWLEtGnT0tPTra2t09PTBw8ezG7BSB1hjCJNcevWLZKnmZmZT548IQd1dXWbmppsbGxOnz5tb2/PboVITWGMIo0jkUiysrKEQqFQKHzy5Imbm9vKlSsHDhzIdl1IXWGMIo1WX19vRN/apUgzYIwihFCX4PR7hBDqEoxRhBDqEoxRhBDqEoxRhBDqkv8H0Rk0hGXwqLEAAAHXelRYdHJka2l0UEtMIHJka2l0IDIwMjAuMDkuMQAAeJx7v2/tPQYg4AFiRgYIkARiGSBuYGRjSACJM0NoJiY47aABpJlZ2CE0M4yPoDPA4oyYDIQWTCXohmJogZoOdgUzI4arcNFQfdxAHzIyMTAxA9kZTCysCaxsCmzsGUzsHAkcnBlMnFwJXNwZTKw8GUw8vAm8fBlMfPwJ/AIZTAKCCYJCDELCDMIiDCKiDKJiDGLiDIISGUwSzAncbAkSvAniQgwiLGzMErw8rCxs7BycXNxsbHz8AoISvGzCIqJi4kLisLBlkPz9j9Hh//1aBxDH0qXMPrLbCsz+GLTaXsZmtz2IHXen087O+5IdiD1t6rT9hh3+YPEdO/gPLNQs3wdi89nnHFieoARWs0e76cBqtXNgNS/yNhz4tZoBbOa9pZcO+MWWgsVXzjx2YM/6P2C9dWEzDhTsStkPYs92eL//1P4bYPb28zf2df5gPQBi6yc77L3yNAnMnsK52b6yrgbM3v5J1uGypiWY7XuL1YG/cgdYr6WHlwOX/XGw+R7pkx3EmbzA4nbztjqwLmwDu5Nz0kqH6023we7ZM7fMQSCKHezO8/pGDpe5OsDi1/1V7Y+5t4L1igEA0xZ9vk3Xz6EAAAG5elRYdE1PTCByZGtpdCAyMDIwLjA5LjEAAHicpZU9bhwxDIX7OYUuEIH/EutsKsMuXOQO6XN/hBrN0msgBmx6MFh8AqWHpydq9mjreb09/fnb8qHbcTTSRrM1+O/r7u03AcBxzu/ASIu4u9lcBD2q0H62jyQe32OvcCPYeo5SVMHOPHWTGJW9kDlvEoRRUvmBHYfJXjvfq7x8XoU6idqicCUoNS/czfHaEeq0mop0sDF3uoZQVNGuyHp1DoLWVKzrpCtT11lWITDe6Yrzo8oXzkj6MLjOaL5X+VK/THPaFO1XPOnYB50OonPAmasqIAN256gol3JZt9Hn2CouSrXbSHGTh+8dRbpU87LSOG9PpCuCRS/c0c6bHAnx9KKK9GgX2SpjOtZUtNtQu3+vyKsq5NsB9oFSVOE+dXcsdSyrUKQhtL3AN77eJjr2SYNg6QZEIHT+rkEQ5yBIchCkZ30NgiwrQSMHQTOnBXlWvCHcK0GI90rQ6lG6q61tXNO4YToIwnQQhOkgCEdWRsOZldkwHQStf909CKJ0EESZQRBlBkEkD2qUDha9Veghg6i8uY7V6eC5tV8vt+MfN8VRZlTaCHwAAAEzelRYdFNNSUxFUyByZGtpdCAyMDIwLjA5LjEAAHicJZHLjcMwDERb2WMCyIL4J+GjS9E9FaT4DLXyQcbDcEiOnufZ9Nmva/Pe+7P5vTd/8PN65Okj7038933xXEJUQ2a557gXbqrIwbNIw8dNUyQ1Bk11jmgJe7HVWFNphYz7oknhqiBJy4OBeLKa27ggVxLUXTK9SAwqsrSj0gl5KsydlqHbZdOgUfRfIAA+LTm6qizV/hEvZ29vLTHO4xTeDmBZonGmSjQEgnuJdCUKWBQI9iVcBy31wjcwoOlxw86V2AykQDApY/so7ULYUzZZSMjaXbUjvGWSLz8jSBYZktEJIgdF1oLBbdPDLEbHyNWD2uSqlQghCPu0URpCRwZEmtG9EBL1G2DvlH4DV4vTfCmxjff3B38jZsctMtPuAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<rdkit.Chem.rdchem.Mol at 0x7fe63ca1f670>"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fused_smile = fuse_on_atom_mapping(fragment_smile)\n",
    "new_mol = to_mol(fused_smile)\n",
    "new_mol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "def murcko_scaffold(smile, generic=False):\n",
    "    '''\n",
    "    murcko_scaffold - convert smile to murcko scaffold.\n",
    "    \n",
    "    If generic, all atoms are converted to carbon\n",
    "    '''\n",
    "    mol = to_mol(smile)\n",
    "    scaffold = MurckoScaffold.GetScaffoldForMol(mol)\n",
    "    \n",
    "    if generic:\n",
    "        scaffold = MurckoScaffold.MakeScaffoldGeneric(scaffold)\n",
    "        \n",
    "    return to_smile(scaffold)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA4QAAAEsCAIAAAAU/OrGAAAABmJLR0QA/wD/AP+gvaeTAAAgAElEQVR4nO3de1zUdb7H8e8goIAp3jUveEFFUTQUb1Rq0tpJslql2naptgue3U7oOe42uD0K261taLdtdO24wyMfj8hOe8Ly9KCbNeYlLVPxgpoXvOH9CnhhUEDme/74siPCAAPMzHcGXs//HH7MfBj58nvP7/v5fn8GKaUAAAAAdAjQXQAAAABaL8IoAAAAtCGMAgAAQBvCKAAAALQhjAIAAEAbwigAAAC0IYwCAABAG8IoAAAAtCGMAgAAQBvCKAAAALQhjAIAAEAbwigAAAC0IYwCAABAG8IoAAAAtCGMAgAAQBvCKAAAALQhjAIAAEAbwigAAAC0IYwCAABAG8IoAAAAtCGMAgAAQBvCKAAAALQhjAIAAEAbwigAAAC0IYwCAABAG8IoAAAAtCGMAgAAQBvCKAAAALQhjAIAAEAbwigAAAC0IYwCAABAG8IoAAAAtCGMAgAAQBvCKAAAALQhjAIAAEAbwigAAAC0IYwCAABAG8IoAAAAtCGMAgAAQBvCKAAAALQhjAIAAEAbwigAAAC0IYwCAABAG8IoAAAAtCGMAgAAQBvCKAAAALQhjAIAAEAbwigAAAC0IYwCAABAG8IoAAAAtCGMAgAAQBvCKAAAALQhjAIAAEAbwigAAAC0IYwCAABAG8IoAAAAtCGMAgAAQBvCKAAAALQhjAIAAEAbwigAAAC0IYwCAABAG8IoAAAAtCGMAgAAQBvCKAAAALQhjAIAAEAbwigAAAC0IYwCAABAG8IoAAAAtCGMAgAAQBvCKAAAALQhjAIAAEAbwigAAAC0IYwCAABAG8IoAAAAtCGMAgAAQBvCKAAAALQhjAIAAEAbwigAAAC0IYwCAABAG8IoAAAAtCGMAgAAQBvCKAAAALQhjAIAAEAbwigAAAC0IYwCAABAG8IoAAAAtCGMAgAAQBvCKAAAALQhjAIAAEAbwigAAAC0IYwCAABAG8IoAAAAtCGMAgAAQBvCKAAAALQhjAIAAEAbwigAAAC0IYwCAABAG8IoAAAAtCGMAgAAQBvCKAAAALQhjAIAAEAbwigAAAC0IYwCAABAG8IoAAAAtCGMAgAAQBvCKAAAALQhjAIAAEAbwigAAAC0IYwCAABAG8IoAAAAtAnUXUBrtG/fviNHjpSVlT3wwANBQUG6ywEAANCGK6NeVVBQMGfOnJEjRz7++OOzZs2KjY3dvHmz7qIAAAC0IYx6ybFjx5599tnBgwdnZma2adNm2rRpAwcO3LNnT3x8/Pz580tLS3UXCAAAoAFh1OPOnz+flpYWFRW1bNkyu92elJS0d+/elStX/vTTT+np6QEBAX/7299GjBhhtVp1VwoAAOBtBiml7hparIsXL/71r39dvHjxtWvXAgICZs2a9dprrw0ZMqT6MXl5ec8++2xubq4QIikpaenSpV26dNFULwAAgLcRRj2iqKho8eLFb7/99pUrVwwGw4wZM1577bVRo0Y5PfjGjRvvvPPOSy+9ZLPZevTo8eabbz7xxBNeLthPlZSUBAcHBwcH6y4EAAA0EWHUza5evfrf//3fJpPp0qVLQoiEhISMjIzY2NgGv/Hw4cMpKSlr1qwRQiQmJi5durRPnz4eL9dv7dmz549//OOePXvKysqSkpJeffXVtm3b6i4KAAA0GmHUbWw225IlSzIyMoqLi4UQCQkJf/7zn+Pi4lx/BinlP/7xj7S0tCtXroSHh7/zzspf/GKqweCxiv3T3r17Fy5c+PHHH0sp27ZtW15eLqUcOXLku+++O27cON3VAQCAxmEBkxuUlZVlZmZGRkampaUVFxfHx8evXbvWarXWTqIXLlyYP3/+999/7/R5DAbDb37zm/379//85z8PCAicOzd68mRx4IDnfwA/oTbGiomJWbFiRVBQUEpKytGjR7/77ruoqKjdu3dPnDhxzpw5JSUlussEAACNIdEMZWVlFovl9ttvV2/mxIkTrVar0yMLCwsXLFjQvn17IcTUqVMbfOaVK4937y6FkKGh8q9/lTduuLt0v3Ls2LGUlJTAwEAhhIqhJ0+edHz12rVr6enpqnN0wIABX3/9tcZSgZuOHJEvvywfe0w+/bRcvlxWVOguqKaUFPneezf/uW6dzMjQVw2A1oow2kTl5eVZWVkDBw5UMTQmJiY7O9vpkVevXjWZTOHh4erIhISE3NxcV16iqEimpEiDQQohR4+Wrn1TS3PixInU1NR27dqpGJqcnHzo0CGnR+bl5Y0dO1a9yUlJSRcuXPByqcAt1q+XYWHyl7+Uy5bJjAwZESHvvdfX8qgQsl07mZ9f9c933pGTJmktCECrRBhttMrKyuzs7MjISJV7oqOjs7Oz7XZ77SNLSkrMZnP37t3VkfHx8evXr2/sy331lYyIkELIwEBpNMpr19zxM/iDCxcuGI3GkJAQIURAQEBSUlK+45xZh4qKCrPZHBYWJoTo0aNHVlaWd0oFaqqslJGR8ve/v/nIhQuyUye5dKm+mpwQQj70kExIqPonYRSAFoTRRlAxdOjQoSpcDhs2LCsr64azGXQ1fd+zZ09HDF2zZk2TX9dmk0ajbNNGCiEjI2Uznsk/XLx4MT09vUOHDkIIg8GQmJi4c+dO17/90KFD06ZNU+98YmLiiRMnPFcq4NyhQ1IIefz4LQ++8IJ8+GFNBTknhNyyRd5+u/zwQykJowA0IYy6xG635+Tk3HHHHSri9O/f32Kx1BNDe/furY6cMGFCTk6OW2r44Qc5fLgUQhoMMiVFXr4s7XZZVCSvX795zLVr8upVt7yaHleuXDGZTB07dnS0NGzbtq3+b6moqLh48WKNB+12e1ZWVufOnYUQHTt2NJvNlZWVHqsaGtjt9uzs7LFjxz722GOpqamXLl3SXdGtvvlGBgfLGr91ixbJmBhNBTknhNy7V77/vuzZUxYXE0bRCL4+BqWUUhYWFqampj733HPR0dFZWVlO5zC127x588SJE+fPn5+YmFhXH1qLRxht2Pr16x0xNCIi4t13361w1vhVo4t05MiRdU3fN9n16/KVV2RwsBRC9u0rP/tMCiHvvffmAW++ecs//UhJSYnJZOrUqZMjhm7ZssWVb/zzn//cvXt3pzPyp0+f/vnPf66e8K677tq/f7+7q4YeP/744/jx49X/bEBAgBCie/fumZmZPvSR47vvZECALCu75UGTSU6YoKkg51QYtdvl5MnyhRcIo3CV74/B8vJys9mszilq8asQYsqUKY2aZ/O0Y8eOPfroowaDQQgRGhoqhGjXrt2CBQuu+vVVpSYhjDZA3VlejTSTyXTNWc+mmr4fPHhwg12kbrF7txw/Xgohs7OlELJr16opNumfYdRms5nN5h49ejhaGtauXev6t99///3qGx944IHqS+wdcnJy1HYHISEh6enpal9S+KlTp06lpKSok1+vXr0sFktubu7dd9+tfgfuuOOOJrRlu9+VK/L0aSmErHHa++UvZXJy1QFa2Wxy6VJpt1eFUSnlnj0yJES+8IKcNEleuCCzsqRPXkKCfn4xBlevXj1ixAhV0rRp03bu3JmVlaXWbwQEBCQnJ589e1ZvhTabzWQyqQ12QkJCjEbjwYMHa7yxvpPsvYAw2oB169YJIaKiopzGUDVPERUVpX7po6Ki6uoida8bN+QXX8iSEimEXLSoaopN+lsYrb0x1urVqxv7JGpGvkuXLkKIDh06OJ2RLy4uTklJUZ8+R40atXXrVjf9BPCesrIys9msOomDg4NTU1MvX77s+GpOTs6AAQMcjcJHjhzRU+XBgzIpScbGyspKOXWqfPBB6fjws22bDA6Wq1bJNWtkeLg0mW7psPGinBzZv78UQi5ffjOMSil//3sZHCwnTZK/+Y0UQo4ZIzds0FIgfJRfjMH8/PykpCRVRmRkZPVdboqLi41Go7pXX3h4uMlkuq5tDOb079/f8V4dPXrU8aXc3Nz4+Hj1pTFjxmxoNYOQMNqAzz77TAgxY8aM2l+yWq3Vp+8tFovT6XvPUWH02DGZkCB/+1sp/SeMqpYGx1+uejbGctGZM2dmz57tuLa6b9++2sesWrUqIiJCTdkYjUanny7gm3JychwNMImJiYcPH659TGlpqclkuu222xxXGq548wLk5cvyd7+r6qHp0EHu2SOPHpVDhsjoaPn88/LRR2VYmHzpJSmlnDtXCiGFkFFR8quvvFehlLm5Mj6+6sXHjJGbNt0SRm02GREhJ02S//M/snfvqvb0X/1KOptvQKvj+2OwpKQkPT1dZc2wsLD09HSnWfPAgQOJiYnqBxkyZMhnn33mtQqllNu2bbvrrrvUq8fGxn733Xe1j1EXudTZymAwJCUlFRQUeLNILQijDfjwww+FEI8++miNx9977z31+9SvX7/MzEwtk7+OMLpvn2zXTm7e7AdhtMbGWCNGjHBjS0NOTo5aOtauXbv09PSyGh17UtpsNqPR2KZNGyHEoEGDmrPFAbxj79699913n2Pm4auG0tupU6eSk5PVVfDbb7/dG1NdlZUyK0v27FkV35KT5ZkzVV8qK5Offir/8he5ZIms/gHJapXR0VWpMCFB7tnj2QqlvHhRpqZW7cjRpYs0m6vuo2G1ypKSm4fl50vVqm2zyfR0GRJSdd+N9HRZWurpGuGjfH8MVlZWZmVlqV4vNQt/xjEG62C1WqOjo9UPlZCQsMcbY/BiamqqOvt06dLFbDbXP4lqs9nS09PV5oahoaHp6emlLXoQEkYbkJmZKYR49tlnazx+7dq1uLi4JUuW1E48XuMIo1LKBQvkhAkyI8N3w6iKoUOGDFHjX22M5fY/UtVn5GNiYpyugvrhhx/UnyGDwZCSklJ9pgm+Qy2DVX+7O3fubDabXZ952LJly8SJE9VvWlxc3Pfff++pKjdvlhMmVMXKcePkpk2ufmN5uTSbZceOUggZFCRTU6VnFiM353VOnJDJyVX33ejTh0bSVscvxuDmzZsnTJigXmjcuHGbXB6DaoWTuh9NUFBQampqsWp3czf1QmqXGPVCru88cOLECUey79Onj89uCNB8hNEGvPXWW0KI//zP/9RdiBPVw6iaYhszxnfDaFZWlvp7MXjw4A8++MCjn5XXr1+vUm9gYGBqampJ9Ys/Ukopy8vLTSaTmtDp1avXypUrPVcMGquiosJisXTt2lX9D6akpNR/Py2nU4GqmVjt9aumuo7X2PWzmU6evJnUevduYlKr64qlm1itVfvBqSuwP/3UlCdZt06OGlX1JJMnyx073FggfFRjx6BTnh6D1YNa7969mxbUagdu9676sFqtw4cPd1yC/alJg3DdunWjR49WTzJ58uQdLXEQEkYbsHDhQiHEyy+/rLsQJ6qHUSnl//1fzZ2efMeRI0c6derUsWPHZcuWeaeztrS01DEjP3DgQKdLo3bv3u34SJ2UlHT+/HkvFIb61VgGu3v37vqPr6ysjIuLS0hI2OtofqxGtZGp28mqNjI39ArbbNJkku3bSyFkSIg0Gpu7u+/27fLuu6vi3ujR0h2Lkffvl/ffX/WUQ4fKL75o1rOpToTu3aUQMiBAJidL3WuR4UGNHYP188QYrL4UPTQ01Gg0NnMvpB07djg2BBg2bNiqVauaWaGUcv/+/Y7NXoYOHfpF8wZh7VYE7RsCuBdhtAG/+93vhBBvvvmm7kKcKC2VAwfKU6duPvL44/KJJ/QVVLcdO3aoeXMvv+727dtjY2PV5/Lk5OTCwsIaB9y4cePtt99WdxDt2rXrBx984OUK4VB9GezgwYNdXNO2e/dutWCibdu2L774otOrpMeOHUtOTlbP3Ldv3ybfJ9Zut3/44Ycn7ruvqj30scdq3mOpOT76SN35124wvPLb3zZ5yUJhYeELL7wwefI+IWTnznLxYtmET39OFxkXF8t582RQkPrQ+/7bb7/NRmktTNPGoCvcNQZlvUvRm8ldGwIUFRUZjcbg4GAhRKdOnUwmk7va+XxnQwC3I4w2YM6cOUKIpT52R+ka3ntP/u//1txg26ds2LBBCBEfH+/9l64xI//xxx/XPubo0aM/+9nPhBAzZ85swmwUmunq1auOZbDt27evaxlsXS5cuODKyoA1a9bExMSoM83UqVPz8vIaVaRjGeyUvn0rJ0zwyL5HpaXy1Vc/nDJFNGkxckVFxZIlS9Q2Zz16jHjhhRu17k3mkqIi2auXTE11vh3qvn3yyScLDYYAtZzlyy+/bMprwMc0cwy6yF1jUNS9FL2Z1PZV6vOt2r6qUWNQXb/s1q2b4/rluXPn3F5k9Q0B3PuZQSPCaAMef/xxIcTy5ct1F1Kfdu2kENJm011H3b788kshxH333aergIMHD06ZMsXxkfdU9evJUkop7Xa7OsBisWipsHVy49zTtm3b7rzzzvpPVE07VTR2GWwzNW0xcjNP89W9997NRUsffOC8Fdb7i5HhIV6e/23aGHTxA6e7VN/YX8sYdEULG4OE0QbMnDlTCPHpp5/qLqROFRVSCNmmjU8vdP3oo4+EELNnz9ZYQ2Vl5aJFi1SbUZcuXb799tsaBzz22GNCiA8d97OCh1W/o2CjlsHWw5UpPNcn0ZqzDLaZXF+MfPz4cXdNgFZ7dTlpUlXLaVycdPriNRYjN22NC/TyxBh0RWPHoNpp3/tjcNKkSerNGTt2rJfHoCta0hgkjDZg6tSpQojawcV3FBdLIWTHjrrrqNeyZcuEEL/+9a8djxQVFfXq1Ss2NtbLlZw6derBBx/s0KFD7XWdM2bMEEJ4eQ/k1smj+5W4uLihweUFVqt12LBhjgsPTVsG2xxq7+u+ffuKfy1GPuZYriil9NDyrJuvLrOzZd++Vf2xSUny1hev4unFyPAQX9gzyF/GYL9+/bSMQRe1jDFIGG3A2LFjhRCbN2/WXUiVf/5T3nOPrN7Cevx41YSaLzObzUKI1NRUxyPHjh1THyK11JOfn1/7QbWacu3atV4vpxVx+zLYuri47YvTjVfcuwy2mZye7Ty+cdW/XLkijUbZtq0UQt52m3znnR1OWwl37NgxefJk9Y65azEyPMRrY9BFjEG32Lt37/Tp0/13DBJGG6DuO+901xgtXn9dCiHT0m4+8tNPVXcW9GV/+tOfhBB/+MMfHI/s2bNHCDF8+HCNVdWg1t3n5ubqLqTF+uabb/r06aP+dv/iF7/wwt9uVzbEvn79ekZGhuOO2+PGjQsMDFTXGBYvXuzle/zW5dChQw899JBjHtBxuWjixIlO7+zgXsePy+Rk2bdvRbt24fXMQjruYSiESEpKsvlyG3tr5f0x6ArGoLv47xgkjDZA3V7yxIkTugupsmCBFEK+/vrNRzZvrurr8mVGo1EI8cYbbzge2bRpk8oHGquqYfDgwUKIAwcO6C6kZbp8+fIf/vAHg8EQGxu7wRNL0evg4voMtUopICCgX79+6jAf3HdWrZBo06ZN+/bte/Xq5Y2bnVazdm1eg9tPqsXIYWFhQ4cO9btrMy2erjHoIsagW/jpGCSMNkB9VvPQXcKa4D/+QwohFy+++cjq1VIIec89+mpywW9/+1shxN///nfHI1arVZ3SNFZVg5pwqb3QHm6Rn5+vZsy1/O2uvnONmmhzOt385JNPCiHmzZvn/QpdVFFRoRZUNXkj0uZwcSm06hGfOXOm9ytEPfSOQRcxBt3C78ZggEDdpJQ2m00IodprfMHVq0IIUb0c9chtt+mpx0UlJSVCiNuqVXn16tUaj2hXu0i4XUhIiNowxcvat2+/cOHCXbt2zZgxw2azvfrqqzExMatXr65xmPpAoi6j+qbAwMCgoCAhRGhoqPdfPSAg4Iknnjhw4IDRaAwMDFy+fHlUVFRGRkZ5eXn1w9Rep/BNusagixiDbuF3Y9B3fyN9wbVr1yorK0NCQlT/ii8oKRHi1uhZO576oNrR09fCqN1uLy0tNRgM6m5MaJGGDBny+eefq7sd5ufnX7p0SXdFfkntxbN9+/Z77723uLg4LS3tk08+0V0UAD9GGK2PrwUm4ew6aO146oPUO1n9AnPtR/Sy2Wx2uz0sLMyXrxnALaZNm7Z9+/Z//vOfs2fP1l2LH4uOjv7mm28+/fTT2bNnP/roo7rLAeDHOO/Wx2fDaAuYpve1OXEf/L9uDebNmzd79uxTp055+XWDgoLUPQ5ahqVLl86ePXvNmjXef+kHH3xwxYoVfIQD0Bz8BamPr129E86ip19P0/vOe+tr9bQSX3/99SeffHLlyhXdhfi33NzcTz755OjRo7oLAYCm8JVWSN/ka1fvhBCDBv1vly4dO3SYJERH9UhQ0LlRo9p27RogRAe9tdWjdtTztffW1+oBAKCV4MpofXxw6nbjxv9Yv/7+9u0rHI8cP/5SXl6n4OBsjVU1yPdX0/taPQAAtBKE0fr44NSt768EcqquMOo7ZftaPQAAtBKE0fr42tRteXl5eXl5YGCgukmu4mtF1lZWVlZeXh4cHBwcHOx40NfK9rV6AABoJQij9fG1qVungcnXiqzN6UVHXyvb1+oBAKCVIIzWx9cCitN6fH9+2S8ytO+/jQAAtEiE0fqoFOU7AcVpgPP9+WW/yNC+/zYCANAiaQujZ8+ePXLkyP79+3UV4Iq6rt7Z7fbdu3frqsfH57tr84sM7ftvIwAALZKGMFpYWLhw4cIhQ4bMnj07Ojp6zpw5Kpf4oLqu3i1dunT06NHer9xpgPO1y7e1Oa3Q18Kfr12pBQCglfBqGL106dLLL788YMCAV199taSkpKKiIiAgIDMzMyYm5ptvvvFmJS6q6+pdYWGhqnz06NHffvut1+rp1q3bM888c++99zoesdvtpaWlAQEBoaGhXiujsWrnzsrKymvXrgUEBISEhOir6xa+dqUWAIBWwkthtKSkJCMjY+DAga+99trVq1cTEhK2bNmye/fubdu2xcXFHT16dPr06Y888sjFixe9U48rTp8+vXPnTiFE7Rtnv/LKK7m5uWPHjj18+HBCQsIjjzxSWFjohZJGjx797rvvvvjii45HSkpKpJRhYWG+fG/oeu4FajAYtJV1K1+7UgsAQCvh8QRTWlq6aNGiyMjItLS04uLi+Pj4devWWa3WsWPHCiFiYmJ++OEHs9kcFha2YsWKESNGvP/++54uqUEXL15MS0uLjIwsKCgICwubO3du7Rn5UaNGbdq0yWw2h4aGqspXrFjh/VL9IkLVnqb3wcuQTNMDAKCFB8NoeXl5ZmZmZGTkvHnzzp07N3HixNWrV2/cuHHy5MnVDwsMDJw7d+6uXbumTZt27ty5J598MjEx8cSJE54rrB5FRUULFizo379/RkbG9evXH3744aeeeiowMDAzM3PkyJE1egkcld9zzz1nz5595JFHHnjggZMnT3qzYN9vGBXOEnOPHj0OHjy4atUqfUXV5IP5GACA1sAjYbSiouL999+PioqaM2fOmTNnxo0bl5OT88MPP0ybNq2ubxk4cKDVas3KyurcufMXX3wxcuTIRYsW2e12T5TnlGokGDRokMlkstlsCQkJW7duXbly5ZIlS7Zu3RoXF1dQUOC0l2DQoEGrV6+2WCwdOnT4/PPPR4wYkZmZKaX0Qs0nT558/fXXO3TocPHixR9//NELr9g0tRNzUFBQZGTkiBEj9BVVk19cYwYAoOVxcxi12+0rVqwYPnz4k08+efTo0REjRmRnZ//4448PPPBAg99rMBieeOKJn376adasWZcvX543b97kyZO9sPeTzWbLyMiIiIhIS0u7dOlSQkLC5s2brVbrmDFj1AEN9hIYDIaUlJT9+/c//PDDly9fnjNnzpQpU/Lz8z1X89mzZ+fOnRsZGbl8+fLS0tKioqL4+Ph58+bZbDbPvWiT+VrOc/pRgWl6AAC0cFsYVTF02LBhjzzyyKFDh4YNG5aVlZWXl5eUlNSoRSo9e/b8+OOPc3JyevfuvXHjxtjY2IULF1ZUVLirzurKyspUI0FaWprKc2vWrLFarePGjatxpCu9BL169Vq5cmV2dna3bt2+++670aNHZ2RkVFZWurfmwsLCtLS0QYMGLV68uKKiIikpadu2benp6YGBgYsWLRoxYsTXX3/t3ldsvrNnzwohfGHhfEFBwZw5c/793/+99peYpgcAQA/ZbHa7PScnZ/To0eoJ+/fvb7FYbty40cynLS4uTklJUUF21KhRW7dubX6pDmVlZRaL5fbbb1c1T5gwIScnx5VvtNvtqpdACNGxY0ez2VxZWVnjmKKiopSUFPXMo0eP3rZtm1tqvnLlislk6tixo3rmhISE7du3O76al5cXFxenvpSUlHThwgW3vGgzHT9+PCUlpU2bNn369Bk4cOCmTZt0VVJQUPDMM88EBgYKIUJDQwsLC6t/1W63t2nTRghRUVGhq8IWT80VREZGVn8wKipKCLF3715dVdVgNBqFEG+88YbuQurTtWtXIcT58+cdjzz99NNCiHfffVdjVdV9+umnQoiZM2fqLgS3cDoGfY2fjkFf43djsLlXRlevXh0XFzdz5sydO3f269fPYrEcPHhQ5Y9mPnN4eLjFYlm7du3gwYPz8vImTpw4d+7c5k9Dq37WYcOGzZkz5/Tp0zExMdnZ2Zs2bXKlkUC41kvQqVMni8Xy1VdfRURE7Ny5c/z48WlpaWVlZU2uWfWz9uvXLy0t7fLly6qf1Wq13nHHHY5jfG1fglOnTj3//PORkZGZmZkGg6GsrOzIkSNaegnOnz+flpYWFRW1bNkyu92elJS0a9cu9XHCobS0tLKyMjQ0VKVVAADgPU2OsdWns/v06WM2m69fv+7GmOxgs9mMRqNKt4MGDfr222+b9jyVlZXZ2dmDBw9WNUdHR2dnZ9vt9iYXpnoJhBAhISHp6enl5eX1VB4ZGbl27drGvoTNZjObzT169FA1q42x6v+Ww4cPOxaKzZgx4/jx44190Wa6cOGC0WhUk/IBAQFJSUkHDhy4dhUE7kIAAA9ISURBVO1aenp6cHCwEKJ///6rVq3SVYnTIzdu3CiECA8PLysr80JhrRNXRt2FK6NoGq6MugtXRt2uKWF048aNU6dOVXGnW7duJpOptLTU7ZXVsH37dnUh0GAwJCcn15hmrZ+KoUOHDlU1R0VFZWVlNb+RQLrWS7Bx48Zhw4apPPTOO++4+MyqkaBXr16q5kmTJrmewl3pJfCEwsLC9PT0Dh06qP+mxMTEnTt3Vj/Aa70EDVbiUFBQkJKSEhgYGBISYjAYBg8e3GDcR9MQRt2FMIqmIYy6C2HU7RoXRjdt2pSYmKjCRJcuXdLT069cueKhymorLy83mUxt27YVQvTq1euTTz5x5buqT2erfla39wWuW7dOXXANDAxMTU1Vd0WqXXl4eHh+fn6Dz1ZeXm6xWNQ1VyHE+PHjXexnreHMmTOzZs0SQsQNH37j7rvlvn1NeBIXqX7W8PBwVXNCQkJdnbIVFRWql0AI0b1796ysLPdWcvXqVRcrUf2sQUFBQoigoKCHHnooMjJSfWZITU29evWqewuD0xNhdna2xWIpLi7WVVUNfnoiXL9+vcVi2b9/v8aqqvO7E2ErQRh1F8Ko27kaRrdu3Xr//ferE3ynTp3+9Kc/eTOGVrdnz54JEyY4rq6dO3euriMd93kSQvTt29dzjQTStV6CBs+46gruoEGDVM1qY6zmNBJIKT/66KNdEyZIIWRIiMzIkG5foHP16sq33+7UqZOq+b777tuyZUuD36Ruo+reXoKSkhKTyeSoRN1y1umR58+fNxqN7dq1c0zfHzx4UEqppZeg9fCpE+H+/fuffvrp2lM6nAgbpbCw8IUXXthX64Ou350IWwmfGoN1YQy6hd+NQVfD6Pz584UQ7du3NxqNRUVFHq2pQXa73WKxqF141GqhGoltw4YNjvs8de/e3WQyXbt2zQuFVe8lSElJuXz5sovfqGLokCFDVM3Dhw9vfgy9qbhYpqRIg0EKIUeNku7al+D6dWmxyJ49L8bGCiHi4+Mb1RTrxl6C69evWyyWnj17qndPbdHl9EjVRRoaGuqIobUvJu3atcsH9yVoAXzkRFhcXPxf//Vf6or466+/XuOrnAhdVFFR8fe//12N3xkzZtT4qt+dCFsJHxmD9WMMuoXfjUFXw+j58+cXLFjgUyfmo0ePTp8+3XFBrqCgQEr5/fff33PPPerBrl27eqeftbrG9hKojbFGjRqlah4wYIBbNsZyYt06OXiwFEIGBsrUVFmrl6ARrl+XixbJnj2lEFIIGR+f29Q+S0cvgRDizjvvrH2JpX6ub9FVu4t0x44ddT2tp3sJWiftJ8LKysqsrCy1HDAgICA5Ofns2bM1juFE6Ipvv/125MiRatDdc889eXl5NQ7wuxNhK6F9DLqCMegWfjcG3bDPqF7vvfee+nQeFhbm+PvYpUuXN954o3bvpte42EtgtVpjY2PVYREREZ7oZ72FzSaNRtmmjRRCDhokm7AvQXm5zMqSAwdWxdCYGJmd3fy6HPsStGvXLj093ZUl7eXl5VlZWQMHDlTvntqiy+mRtbtIc3NzXanKE70ErZk6EXbu3FnLO7lu3TrHXsjjx4//8ccfax9z/fp1NaPiFydCLW0kBw8eTEpKUm9jZGSk00Fnt9t///vf+9eJsJXQOwZdwRh0C38cg34fRqWU586dS05ODggIiIiIUI0EvrAeorKyskYvQfWvWq1Wx1ywRzfGcmL7dnnHHVIIaTDI5GTp4r4ElZUyO1tGRlbF0OhomZ0t3dVIcOu+BDExMQ3e4+D06dNqtr2eLbpKSkrMZnP37t0d0/fr169vVFXVewk6dOjgtX0JWqSKioq33npLCBESEmI0Gr22ROzEiRPJycnqV6tPnz5ZWVlOf1tycnJUu/b06dMvXbrkndqaZv369UFBQQaDISkp6dixY9550ZKSkvT0dNVpHRYWlp6e7rT3afv27XfffbeaF1KzVfAdusagixiDbuGnY7AlhFFl9+7deXl52vtZazh69OjPfvaz6r0EGzZsmDJlinpEbYzlnX7WW5SXS5NJtm0rhZC9eklHL8HmzfL11+WCBXLZMukI9CqGDh1aFUOHDZNZWdITjQQu7EtQ3dtvv/3RRx85TYc1NsaKj49v8va0UsozZ87Mnj3b8VSN7SWAw8mTJx25sHfv3nXlQnex2Wzp6elqo9nQ0NC6zr779u37t3/7N/X/O3To0C+//NJzJbmF+rlULgwNDa0rF7qL+kimerLVznpnzpypfdjFixdTU1PVIs4uXbqYzWaPtBuhebw8Bl3EGHQLvx6DLSeM+jJHL4HqJVUx9C9/+YvNZtNZVn6+nDxZCiF/8xsppUxNlbfdJp97Tr70krzzTtmtm9y2TZ46JaOjq2Lo4MFy+XLp4UuDpaWlzbnHQY2NsVy/0WuDmtBLAKc2b97saGIZN26cJ24Sa7fbs7OzIyIiVH5KSkpyeoWgqKgoNTVV3XarU6dOZrPZj+4He/z48eTkZPU29u3b10OdzVu2bJk4caJ6lbi4uB9++KH2MeXl5WazWd2pOCgoKDU11ccva8ELY9BFjEG3aAFjkDDqJWfPnk1OTn7iiSc6depkNBp95RelslIuXSovX5Y5ObJtW7lnT9Xjdrt85hk5fLisrJSjR8t+/aTF4v5toerWhHscqC5Sx8ZYI0eOdOeOBFLKWr0ErmxiBadcvNjWNLm5uXfeeaf6NYiNjd2wYUPtYyoqKiwWS7du3dRl+JSUFF9ei1CPNWvWOJY/Tpkypa6bOzSBi5fQrFZrdHS0KiAhIWGP428IfJtHx6ArGIPu0jLGIGHUq27cuOG93tBGeeop+fjjtzxy/LgUQu7dK48ckTquArq+L0GNG726eWOsWtavX6824XKllwD1UG2I6r9YtSE2c3ScPn06JSVFXVbv1atXXRtT1FgMvmvXrua8qHZqlwDVG612Cahn92VXlJaWmkwm1e+umgud7ip94MCBGTNmqLdxyJAhn3/+eXNeFFq4fQy6iDHoFi1pDBJGIaWU8s475R//WPPB226TbprjbrL8/HzHlrG19yXw3sZYt6rRS7B69WpPv2IL5soC7QapWSq1dZeapXK6y69bXss3FRcXG41GdcuG8PBwk8nUtE6SnJycAQMGqLcoMTHxyJEjnnst+AhvjgvGoN+9lncQRiGllPKuu5yH0c8+01HNLeq6x4G3N8aqZfv27aoA13sJUJfmXClxLMJV+enQoUO1j6l9BcgXFhy4XXOulDgW4Qoh7rjjDqdbT+i6AgQv8PTVSsagW7TUMUgYhZRSyqeekr/61S2PnDxZNU3vG6rvSzB+/HjH1dB+/fplZmaWl5drqaqx9zhAPZrWQ5aXl6f6GkeMGGG1WmsfoL03zvusVuvw4cMb1UNWWlqqtk7s3r17Zmam000q1q5dq7c3Dp7moT5OxqC7+jhb8BgkjEJKKeXKlbJdO3n48M1HUlNlVJQbdxJ1i+zs7K5du6oPhdo2xqolPz/fsV1XYmLiqVOndFfkxwoLCxu7uvb5559fsmSJ08OqrxquazF4i9SE1bVLly6dP3++08N8Z9UwvKAJY7AejEHXx2A9WvwYJIziX379a9mli3zxRfm3v8nERBkeLp3dpUa7s2fPbt269YMPPvDyjV7rV72XIDw83NFLgKapvu9gVFRUE/Yd9M39FL2s+fsO+uZ+ivACxqBbMAZdRBhFNd98I3/3OzlnjnzrLVnrtt1oUEFBwfTp09Wf72eeeUZ3OX7PlWbQ2tRi8Pbt2wtfvdOMl7nSDFqb2qu1X79+4l97tfranWbgBYxBt2AMNogwCriZ6iVw12b7rZyLy+QdcnJy+vfv7zh3Hj161FuV+jpXlsk7bN26ddKkSergsWPHbty40Wt1wtcwBt2FMVgPwijgfk73ZUSTqQ1EAwICRN0biG7btu2uu+5q7LWHVsWVDURPnTrleKtvv/12i8XidDETWhvGoFswButCGAXgH6rfWmnMmDGOWyv59R2Zva+uTr6ysjKz2axOk8HBwQ1eAEMrxBh0C8ZgbYRRAH6jxk3nZ82a9corr/j1HZl1qXHT+TfffHPgwIGOCcTD1TfWAKphDLoLY7A6wigAP6OWl4aEhAgh1CavCQkJP/30k+66/Ez13R87d+4shBg2bNiqVat01wU/wBh0C8agg0FKKQDA3xQUFPzjH/8YNGhQRESE44YIaKwrV65kZGSMHTv23Llzzz33nJppBVzBGHQLxqAQgjAKAAAAbQJ0FwAAAIDWizAKAAAAbQijAAAA0IYwCgAAAG0IowAAANCGMAoAAABtCKMAAADQhjAKAAAAbQijAAAA0IYwCgAAAG0IowAAANCGMAoAAABtCKMAAADQhjAKAAAAbQijAAAA0IYwCgAAAG0IowAAANCGMAoAAABtCKMAAADQhjAKAAAAbQijAAAA0IYwCgAAAG0IowAAANCGMAoAAABtCKMAAADQhjAKAAAAbQijAAAA0IYwCgAAAG0IowAAANCGMAoAAABtCKMAAADQhjAKAAAAbQijAAAA0IYwCgAAAG0IowAAANCGMAoAAABtCKMAAADQhjAKAAAAbQijAAAA0IYwCgAAAG0IowAAANCGMAoAAABtCKMAAADQhjAKAAAAbQijAAAA0IYwCgAAAG0IowAAANCGMAoAAABtCKMAAADQhjAKAAAAbQijAAAA0IYwCgAAAG0IowAAANCGMAoAAABtCKMAAADQhjAKAAAAbQijAAAA0IYwCgAAAG0IowAAANCGMAoAAABtCKMAAADQhjAKAAAAbQijAAAA0IYwCgAAAG0IowAAANCGMAoAAABtCKMAAADQhjAKAAAAbQijAAAA0IYwCgAAAG0IowAAANCGMAoAAABtCKMAAADQhjAKAAAAbQijAAAA0IYwCgAAAG0IowAAANCGMAoAAABtCKMAAADQhjAKAAAAbQijAAAA0IYwCgAAAG0IowAAANCGMAoAAABtCKMAAADQhjAKAAAAbQijAAAA0IYwCgAAAG0IowAAANCGMAoAAABtCKMAAADQhjAKAAAAbQijAAAA0IYwCgAAAG0IowAAANCGMAoAAABtCKMAAADQhjAKAAAAbQijAAAA0IYwCgAAAG0IowAAANCGMAoAAABtCKMAAADQhjAKAAAAbQijAAAA0IYwCgAAAG0IowAAANCGMAoAAABtCKMAAADQhjAKAAAAbf4f6WKmlq333IwAAAFYelRYdHJka2l0UEtMIHJka2l0IDIwMjAuMDkuMQAAeJx7v2/tPQYg4AFiRgYIEARiISBuYGRjSACJM7M5aABpZhY2hwwQzcwIE+BgANNMMD47hGZGUoimA2IkExMqzYhVBzqDG+hERqYMJibmBGYWBRZWBVa2DCY29gR2jgwmDs4ETi4GLm4Gbh4GTt4MJma+DCY+/gR+gQwmAcYEXtYEESY2RgF+PqBb2dg5OHlZxWHeZRBsNNjvwO+2eT+Io3ZnmsMLfzEwO94vwqH97CIwe/lSdof1e1/tA7HfvPOw/xcyCSz+KP3zvr7bK8DiFnFn953qmWoPYv9ZcWX/k1V/wezsffYHgliXgdkL6l0O3PrkC1Z/7EvPgdvHk8DmKHluPaAfO2UviC3x7/6B09ccweIX1z7fX3anEcy+m8bj4CjSCjbn5IVEhys774DZB2xmOhSZZ4LZYgBgn2CZpDHpCQAAAVh6VFh0TU9MIHJka2l0IDIwMjAuMDkuMQAAeJydlEFuxDAIRfc5BRcYCzAYWHe6qjqVuugduu/9VRynbhatNGMrsj7BefrGOBv08X59+fyCOfi6bUAG5AD45xMR8MGIuPX1WsLEUlyoCGLr77BkFuEJ/kOcn50ixUikU7CoG69RaqkWPrywUaxRuBCFDi+irmsULOYSwwtJPe/o7W5Kd2B61KXyvrfHvXQKEvVv00tzOVFu91OotEYyKOG06IVLRMNB4TBco9SChnXsjY+9PU6RrMs4IyzuutYvF826KA8KGssapWVdRu9m5+jiDcgzckIcCkkWezdvQFQfZ4Q1bPU2KvvRL1mWtbr0P4P74SCIlvol24P3uQep6j73IJXMTCqdQao2g1Q2v0nlM5MqZiag+xuZVL0tj4CApoNUVM/LZAIESH+CVDQdpKLp4FBHRn6NvgI8367bNxRm5vfWxcy5AAAA5XpUWHRTTUlMRVMgcmRraXQgMjAyMC4wOS4xAAB4nCWQzREDIQhGW8lxd8Z1APmdPW4BKcJ7Kkjx+TSe8AkP8HmeyXMe7ykT53jOKednTn59j8t7VUW0i3pYcGq7L+vuxrkYhaiOBbUraW6YaekCNvBOUosJMcti3JOgbAiI1VYpdQ1TkpU3pLLdaJaa5SuLdaxC6cyl26WWGrlRDQQwjXKA0U1yBMAaC0UKT3o06sVM9CcQ7n0sgy3abb1CkYReSmTbM6IgBpHg+o+IBaigZs/hexF3Vgap5DXOJfgrx10qSNv5/QGOf0ckdnHb5AAAAN16VFh0cmRraXRQS0wxIHJka2l0IDIwMjAuMDkuMQAAeJx7v2/tPQYg4AFiRgYI4AViPiBuYGRzyADSzMxYGRogBgsHA5hmgvHxaoEy2CFqmTFluBkYM5gYmRKYmDOYmFkUWFgVWNkymNjYE9g5Mpg4OBM4uYDi3AncPBlMPAwJXKwJIkxsDDzczEyMbGzsHJxcrOIwTzDwBmVVHrjZGbsfxAGx1cuU7EHsh25qB0SsN0LZy/bDxBkYHJYhidsj1C+zh5kD1Ovw5sg5mJkOSOY7IJnvgGw+kt4DML1iAEGFQmrbD+o7AAAA53pUWHRNT0wxIHJka2l0IDIwMjAuMDkuMQAAeJytlMEOgjAMhu97iv8FXNptzPUsnoyYePAdvPv+sUMsmKAxwNKEv/zrRzMKDnVd29P9AVuhdQ4cwQmg2RAR3AIROd2+i77InqsiX3KTKoO8uoQDviGm8UnR2hxTXkYJvpHSU9jHspTCPojQ2l6GirleLn9TsEkvI2XNO4Kdrp5Q5qWUTeYFm8wLfsxLt2BeVn0BW5yuloc+aqKX2N+qiapkjqrGHFXZElV726aqmKNKzBEwvR1VzBM0B3NC/Y8MDr96GRweH3oGjl3rnhZOs54iVoZhAAAAkHpUWHRTTUlMRVMxIHJka2l0IDIwMjAuMDkuMQAAeJx1j8ENhDAMBFvhCVJiZe3EOKKIK8K6LxVQPLkHR6TAy9JotLt2uPv8cW7Hd17cMR1zFLK6AiEmMi1iJWx/lkglK36IqVRLFkBiWbkRENeaci8pqULyN0JvcfQu0lVe8S21KWJhe9g1znpf1b8zhC/HCcNsOEOFeQU2AAAAx3pUWHRyZGtpdFBLTDIgcmRraXQgMjAyMC4wOS4xAAB4nHu/b+09BiDgAWJGBgjgBWI+IG5gZGNIANJMTNhpZkb8fOJpbqDdjEwMTMwMzCwMLKwMrGwMbOwM7BwMHJwMnFwMzNwM3DwgJ3KxMogANfFwMzMxsrGxc3BysYrDXM3AG5RVeeBmZ+x+EAfEVi9TsgexH7qpHRCx3ghlL9sPE2dgcFiGJG6PUL/MHmYOUK/DmyPnYGY6IJnvgGS+A7L5SHoPwPSKAQDfojrfpxFntQAAANl6VFh0TU9MMiByZGtpdCAyMDIwLjA5LjEAAHicrZMxDsIwDEX3nOJfgMhO0jSeKROCgYE7sHN/4VTFbSVYmkRR9L4a/365rkNdj+n6esNWmJwDR3AC6OcWETwDETm9foq+yMiVyJc8pOpBXp8Szvhnsd17F63NMeVjLsEPUmYX9rEcdWEfRKg1y1LRlgVdsqwuLd8I1l3tUOajLl3mBV3mBZ3npekP6NFdLQ/zWYVSNKGUTCgNJpSyCaXRhFIxoSQmBExfocS8sWZLoMSWYKFF8PrSG3C5T+4D3vuzgU4rhqgAAACMelRYdFNNSUxFUzIgcmRraXQgMjAyMC4wOS4xAAB4nHWPMQ7DMAwDv9IxAWzBlGxVRka+pb/I4+sMaQ241Xg4kBRBciOV1+lO4nFu2ST6E0i5SHizaOn4sCJu1XEhldajRIJYVNdBINp7qbPk4g6rrwz/iqt3k6nyjh+pQ7FIx49d66z/q+Z3lvD9fAMliTasm1hgMwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<IPython.core.display.Image object>"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "smile = 'Cc1cc(Oc2nccc(CCC)c2)ccc1'\n",
    "scaffold = murcko_scaffold(smile)\n",
    "scaffold_generic = murcko_scaffold(smile, generic=True)\n",
    "draw_mols(to_mols([smile, scaffold, scaffold_generic]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Structure Enumeration\n",
    "\n",
    "Often it can be useful to enumerate variants of the same core structure. For example, generating every 6 member ring variant with 2 nitrogens. The `StructureEnumerator` class provides a way of enumerating over a core structure defined by a smarts string and a set of user inputs. The structure enumerator can also add wildcard atoms.\n",
    "\n",
    "For examples on using the `StructureEnumerator` class, see the Structure Enumeration tutorial page"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "def add_map_nums(mol):\n",
    "    'Adds map numbers to all atoms in `mol`'\n",
    "    mol = to_mol(mol)\n",
    "    \n",
    "    for i, atom in enumerate(mol.GetAtoms()):\n",
    "        atom.SetAtomMapNum(i+1)\n",
    "    \n",
    "    return mol\n",
    "\n",
    "def check_ring_bonds(smile):\n",
    "    'Looks for SP hybridized atoms in rings'\n",
    "    mol = to_mol(smile)\n",
    "    for atom in mol.GetAtoms():\n",
    "        if atom.GetHybridization()==Chem.rdchem.HybridizationType.SP and atom.IsInRing():\n",
    "            return False\n",
    "        \n",
    "    return True\n",
    "\n",
    "\n",
    "def decorate_smile(smile, num_attachments):\n",
    "    '''\n",
    "    decorate_smiles - adds wildcard atoms to smile based on num_attachments. \n",
    "     If there are N atoms with at lest one implicit hydrogen, N choose num_attachments \n",
    "     combinations are generated\n",
    "    '''\n",
    "    mol = to_mol(smile)\n",
    "\n",
    "    to_decorate = []\n",
    "    for atom in mol.GetAtoms():\n",
    "        if atom.GetNumImplicitHs()>0:\n",
    "            to_decorate.append(atom.GetIdx())\n",
    "\n",
    "    decorated = []\n",
    "    if len(to_decorate)>=num_attachments:\n",
    "        dec_combos = list(itertools.combinations(to_decorate, num_attachments))\n",
    "\n",
    "        for dc in dec_combos:\n",
    "            new_mol = Chem.RWMol(mol)\n",
    "            for atom_id in dc:\n",
    "                new_id = new_mol.AddAtom(Chem.Atom('*'))\n",
    "                new_mol.AddBond(atom_id, new_id, Chem.rdchem.BondType.SINGLE)\n",
    "\n",
    "            new_mol = new_mol.GetMol()\n",
    "            Chem.SanitizeMol(new_mol)\n",
    "            decorated.append(to_smile(new_mol))\n",
    "\n",
    "    return decorated\n",
    "\n",
    "def decorate_smiles(smiles, num_attachments):\n",
    "    \"Decorate all items in `smiles` and cleans results\"\n",
    "    decorated = [decorate_smile(i, num_attachments) for i in smiles]        \n",
    "    decorated = [i for i in decorated if i]\n",
    "    decorated = flatten_list_of_lists(decorated)\n",
    "    decorated = deduplicate_list(decorated)\n",
    "    return decorated\n",
    "\n",
    "def remove_atom(rwmol, atom_idx, add_bond=True):\n",
    "    '''\n",
    "    remove_atom - removes atom based on `atom_idx`. If `add_bond` and the removed\n",
    "     atom is connected to two other heavy atoms, a single bond is formed between neighbors\n",
    "    '''\n",
    "    atom = rwmol.GetAtomWithIdx(atom_idx)\n",
    "    bonds = atom.GetBonds()\n",
    "\n",
    "    if len(bonds)==1:\n",
    "        rwmol.RemoveAtom(atom_idx)\n",
    "    elif len(bonds)==2 and add_bond:\n",
    "        neighbors = atom.GetNeighbors()\n",
    "        n1 = neighbors[0].GetIdx()\n",
    "        n2 = neighbors[1].GetIdx()\n",
    "        rwmol.AddBond(n1, n2, order=Chem.rdchem.BondType.SINGLE)\n",
    "        rwmol.RemoveAtom(atom_idx)\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "    return rwmol\n",
    "\n",
    "def generate_spec_template(mol):\n",
    "    'generates blank atom_spec and bond_spec for mol. returns blank specs and matching smarts'\n",
    "    mol = add_map_nums(mol)\n",
    "    \n",
    "    atom_spec = {atom.GetAtomMapNum():[] for atom in mol.GetAtoms()}\n",
    "    bond_spec = {(bond.GetBeginAtom().GetAtomMapNum(), bond.GetEndAtom().GetAtomMapNum()):[]\n",
    "                 for bond in mol.GetBonds()}\n",
    "    \n",
    "    return atom_spec, bond_spec, to_smart(mol)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "class StructureEnumerator():\n",
    "    '''\n",
    "    StructureEnumerator - class for enumerating molecular structures\n",
    "    \n",
    "    Inputs:\n",
    "    \n",
    "    - `smarts str`: base smarts string to enumerate\n",
    "        \n",
    "    - `atom_spec dict`: dict of the form `{atom_map_num:[allowed_atom_types]}` \n",
    "    where elements in `allowed_atom_types` match keys in `self.atom_types`\n",
    "            \n",
    "    - `bond_spec dict`: dict of the form \n",
    "    `{(bond_start_map_num, bond_end_map_num) : [possible_bond_types]}` \n",
    "    where elements in `possible_bond_types` match keys in `self.bond_types`\n",
    "         \n",
    "    - `max_num int`: max number of combinations to iterate\n",
    "         \n",
    "    - `substitute_bonds Optional[List]`: List of bond types for new bonds \n",
    "    formed from removing atoms. Bond types should match keys in `self.bond_types`. \n",
    "    If None, all keys in `self.bond_types` are used\n",
    "    '''\n",
    "    atomic_types = {\n",
    "                    'H':1,\n",
    "                    'C':6,\n",
    "                    'N':7,\n",
    "                    'O':8,\n",
    "                    'F':9,\n",
    "                    'P':15,\n",
    "                    'S':16,\n",
    "                    'Cl':17,\n",
    "                    'Br':35\n",
    "                    }\n",
    "\n",
    "    bond_types = {'single':Chem.rdchem.BondType.SINGLE,\n",
    "                  'double':Chem.rdchem.BondType.DOUBLE,\n",
    "                  'triple':Chem.rdchem.BondType.TRIPLE,\n",
    "                  'aromatic':Chem.rdchem.BondType.AROMATIC}\n",
    "    \n",
    "    def __init__(self, smarts, atom_spec, bond_spec, max_num=1000000, substitute_bonds=None):\n",
    "        \n",
    "        self.smarts = smarts\n",
    "        self.base_mol = Chem.MolFromSmarts(smarts)\n",
    "        self.atom_spec = atom_spec\n",
    "        self.bond_spec = bond_spec\n",
    "        self.max_num = max_num\n",
    "        self.clean_specs()\n",
    "        self.get_ring_info()\n",
    "        \n",
    "        if substitute_bonds is not None:\n",
    "            self.substitute_bonds = substitute_bonds\n",
    "        else:\n",
    "            self.substitute_bonds = list(self.bond_types.keys())\n",
    "        \n",
    "        key_list = list(atom_spec.keys()) + list(bond_spec.keys())\n",
    "        self.key_to_idx = {key_list[i] : i for i in range(len(key_list))}\n",
    "        \n",
    "        # generate combinations based on atom_spec and bond_spec\n",
    "        self.iterator = itertools.product(*(list(atom_spec.values()) + list(bond_spec.values())))\n",
    "        self.num_combos = np.prod([len(i) for i in list(atom_spec.values()) + list(bond_spec.values())])\n",
    "        \n",
    "        # write combinations into memory\n",
    "        self.combos = []\n",
    "        \n",
    "        for combo in self.iterator:\n",
    "            combo = dict(zip(self.key_to_idx.keys(), combo))\n",
    "            if self.check_combo_ring(combo):\n",
    "                self.combos.append(combo)\n",
    "                \n",
    "            if len(self.combos)>max_num:\n",
    "                break\n",
    "        \n",
    "        # check conditions for atom removal\n",
    "        self.check_removals()\n",
    "        # update combos based on atom removal\n",
    "        self.combos = flatten_list_of_lists([self.check_combo(i) for i in self.combos])\n",
    "        \n",
    "    def filter_combos(self, filter_func):\n",
    "        outputs = maybe_parallel(filter_func, self.combos)\n",
    "        self.combos = [self.combos[i] for i in range(len(self.combos)) if outputs[i]]\n",
    "        \n",
    "    def clean_specs(self):\n",
    "        'cleans empty entries from `atom_spec` and `bond_spec`'\n",
    "        \n",
    "        atom_remove = []\n",
    "        bond_remove = []\n",
    "        for k,v in self.atom_spec.items():\n",
    "            if not v:\n",
    "                atom_remove.append(k)\n",
    "                \n",
    "        for k,v in self.bond_spec.items():\n",
    "            if not v:\n",
    "                bond_remove.append(k)\n",
    "                \n",
    "        for k in atom_remove:\n",
    "            self.atom_spec.pop(k)\n",
    "            \n",
    "            \n",
    "        for k in bond_remove:\n",
    "            self.bond_spec.pop(k)\n",
    "            \n",
    "    def check_combo_ring(self, combo):\n",
    "        \n",
    "        if -1 in combo.values():\n",
    "            return True\n",
    "        \n",
    "        bad_bonds = [['DOUBLE', 'DOUBLE'], ['SINGLE', 'TRIPLE'], ['TRIPLE', 'SINGLE']]\n",
    "        \n",
    "        for ring in self.ring_dicts:\n",
    "            ring = ring.copy()\n",
    "            for k,v in combo.items():\n",
    "                if k in ring.keys():\n",
    "                    ring[k] = v\n",
    "                    \n",
    "            arom_bonds = [v for k,v in ring.items() if v=='aromatic']\n",
    "            if 0<len(arom_bonds)<len(ring.keys()):\n",
    "                # not all bonds in ring are aromatic\n",
    "                return False\n",
    "            \n",
    "            atom_validation = defaultdict(list)\n",
    "            for k,v in ring.items():\n",
    "                a1,a2 = k\n",
    "                atom_validation[a1].append(v)\n",
    "                atom_validation[a2].append(v)\n",
    "                \n",
    "            for k,v in atom_validation.items():\n",
    "                if v in bad_bonds:\n",
    "                    # SP atoms in rings\n",
    "                    return False\n",
    "                \n",
    "        return True\n",
    "            \n",
    "    def get_ring_info(self):\n",
    "        self.base_mol.UpdatePropertyCache()\n",
    "        Chem.GetSymmSSSR(self.base_mol)\n",
    "        ring_info = self.base_mol.GetRingInfo()\n",
    "        self.ring_dicts = []\n",
    "        bonds = list(self.base_mol.GetBonds())\n",
    "        \n",
    "        for ring in ring_info.BondRings():\n",
    "            ring_bonds = [bonds[i] for i in ring]\n",
    "            ring_dict = {(i.GetBeginAtom().GetAtomMapNum(), \n",
    "                          i.GetEndAtom().GetAtomMapNum()): str.lower(i.GetBondType().__str__())\n",
    "                         for i in ring_bonds}\n",
    "            self.ring_dicts.append(ring_dict)\n",
    "        \n",
    "    def check_removals(self):\n",
    "        '''\n",
    "        check_removals\n",
    "        \n",
    "        If atoms removals are possible in the given atom_spec, a new bond may be added. \n",
    "        If the removed atom has the form `R1-atom_to_remove-R2`, the ouput will be `R1-R2`. \n",
    "        This function determines the map numbers of the removed bonds and the added bonds\n",
    "        '''\n",
    "        self.removal_dict = {}\n",
    "        \n",
    "        for map_num, values in self.atom_spec.items():\n",
    "            if -1 in values:\n",
    "                atom = [i for i in self.base_mol.GetAtoms() if i.GetAtomMapNum()==map_num][0]\n",
    "                bonds = atom.GetBonds()\n",
    "                if len(bonds)==2:\n",
    "                    \n",
    "                    neighbors = atom.GetNeighbors()\n",
    "                    n1 = neighbors[0].GetAtomMapNum()\n",
    "                    n2 = neighbors[1].GetAtomMapNum()\n",
    "                    new_bond = (n1,n2)\n",
    "                    removed_bonds = [(n1,map_num), (n2,map_num)]\n",
    "                    \n",
    "                    self.removal_dict[map_num] = {'new_bond':new_bond,\n",
    "                                                  'removed_bonds':removed_bonds}\n",
    "                    \n",
    "    def check_combo(self, combo):\n",
    "        '''\n",
    "        If a combination contains atom removal, the combination is updated to remove \n",
    "        deleted bonds and create a new bond\n",
    "        '''\n",
    "        c_out = []\n",
    "        for removal_num in self.removal_dict.keys():\n",
    "            if combo[removal_num]==-1:\n",
    "                removed_bonds = self.removal_dict[removal_num]['removed_bonds']\n",
    "                for rb in removed_bonds:\n",
    "                    n1, n2 = rb\n",
    "                    if (n1,n2) in combo.keys():\n",
    "                        combo.pop((n1,n2))\n",
    "                    if (n2,n1) in combo.keys():\n",
    "                        combo.pop((n2,n1))\n",
    "                        \n",
    "                added_bond = self.removal_dict[removal_num]['new_bond']\n",
    "                for new_bondtype in self.substitute_bonds:\n",
    "                    new_combo = combo.copy()\n",
    "                    new_combo[added_bond] = new_bondtype\n",
    "                    c_out.append(new_combo)\n",
    "                    \n",
    "        if not c_out:\n",
    "            c_out = [combo]\n",
    "            \n",
    "        return c_out\n",
    "                \n",
    "                    \n",
    "    def _create_mols(self, combos):\n",
    "        \"Creates smiles strings from `combos` and cleans up results\"\n",
    "        \n",
    "        raw_smiles = maybe_parallel(self.create_mol, combos)\n",
    "        clean_smiles = []\n",
    "        \n",
    "        for s in raw_smiles:\n",
    "            if type(s)==str and ':' in s:\n",
    "                s = s.replace(':', '=') # convert incomplete aromatic rings to double bonds\n",
    "                m = to_mol(s)\n",
    "                if m is None:\n",
    "                    s = None\n",
    "                \n",
    "            if s is not None and check_ring_bonds(s):\n",
    "                clean_smiles.append(s)\n",
    "        \n",
    "        smiles = deduplicate_list(clean_smiles)\n",
    "        return smiles\n",
    "    \n",
    "    def create_mols(self, max_num=None):\n",
    "        '''\n",
    "        Creates `max_num` smiles strings from `self.combos` and cleans up results.\n",
    "        \n",
    "        More memory efficient for >1000000 combos\n",
    "        '''\n",
    "        if max_num is None:\n",
    "            max_num = self.max_num\n",
    "            \n",
    "        return self._create_mols(self.combos[:max_num])\n",
    "\n",
    "    \n",
    "    def create_mols_chunks(self, chunksize):\n",
    "        \"Creates mols from `self.combos` in batches of `chunksize`\"\n",
    "        combo_chunks = chunk_list(self.combos, chunksize)\n",
    "        \n",
    "        smiles = []\n",
    "        for i, chunk in enumerate(combo_chunks):\n",
    "            print(i)\n",
    "            smiles_iter = self._create_mols(chunk)\n",
    "            smiles += smiles_iter\n",
    "            \n",
    "        smiles = deduplicate_list(smiles)\n",
    "        return smiles\n",
    "    \n",
    "    def create_mol(self, code):\n",
    "        \"Edits `self.base_mol` depending on `code`\"\n",
    "        new_mol = Chem.RWMol(self.base_mol)\n",
    "        \n",
    "        atoms_to_remove = []\n",
    "        # iterate atoms\n",
    "        for atom in new_mol.GetAtoms():\n",
    "            map_num = atom.GetAtomMapNum()\n",
    "            if map_num in code.keys():\n",
    "                atom_code = code[map_num]\n",
    "                if atom_code == -1:\n",
    "                    atoms_to_remove.append(atom.GetAtomMapNum())\n",
    "                else:\n",
    "                    atom_id = self.atomic_types[atom_code]\n",
    "                    atom.SetAtomicNum(atom_id)\n",
    "                    \n",
    "        for map_num in atoms_to_remove:\n",
    "            atom_idx = [i.GetIdx() for i in new_mol.GetAtoms() if i.GetAtomMapNum()==map_num][0]\n",
    "            remove_atom(new_mol, atom_idx)\n",
    "        \n",
    "        bonds_to_remove = []\n",
    "        # iterate bonds\n",
    "        for bond in new_mol.GetBonds():\n",
    "            start = bond.GetBeginAtom().GetAtomMapNum()\n",
    "            end = bond.GetEndAtom().GetAtomMapNum()\n",
    "            \n",
    "            if (start, end) in code.keys():\n",
    "                bond_code = code[(start, end)]\n",
    "            elif (end, start) in code.keys():\n",
    "                bond_code = code[(end, start)]\n",
    "            else:\n",
    "                bond_code = None\n",
    "                \n",
    "            if bond_code is not None:\n",
    "                if bond_code == -1:\n",
    "                    bonds_to_remove.append((bond.GetBeginAtomIdx(), bond.GetEndAtomIdx()))\n",
    "                else:\n",
    "                    bond.SetBondType(self.bond_types[bond_code])\n",
    "                \n",
    "        for atom in new_mol.GetAtoms():\n",
    "            atom.ClearProp('molAtomMapNumber')\n",
    "            \n",
    "        for to_remove in bonds_to_remove:\n",
    "            new_mol.RemoveBond(to_remove[0], to_remove[1])\n",
    "            \n",
    "        mol = new_mol.GetMol()\n",
    "\n",
    "        try:\n",
    "            Chem.SanitizeMol(mol)\n",
    "            mol = to_mol(to_smile(mol))\n",
    "            output = to_smile(mol)\n",
    "        except:\n",
    "            output = None\n",
    "\n",
    "        return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "def add_one_atom(inputs):\n",
    "    'helper function for `add_atom_combi`'\n",
    "    mol, source_idx, target_atom, bond_type = inputs\n",
    "    new_mol = Chem.RWMol(mol)\n",
    "    \n",
    "    try:\n",
    "        if target_atom == -1:\n",
    "            new_mol.RemoveAtom(source_idx)\n",
    "        elif target_atom == -2:\n",
    "            atom = new_mol.GetAtomWithIdx(source_idx)\n",
    "            bonds = atom.GetBonds()\n",
    "\n",
    "            if len(bonds)==2:\n",
    "                neighbors = atom.GetNeighbors()\n",
    "                n1 = neighbors[0].GetIdx()\n",
    "                n2 = neighbors[1].GetIdx()\n",
    "                new_mol.AddBond(n1, n2, order=Chem.rdchem.BondType.SINGLE)\n",
    "\n",
    "            new_mol.RemoveAtom(source_idx)\n",
    "\n",
    "        else:\n",
    "            target_atom = Chem.Atom(target_atom)\n",
    "            new_idx = new_mol.AddAtom(target_atom)\n",
    "            new_mol.AddBond(source_idx, new_idx, bond_type)  \n",
    "\n",
    "\n",
    "        mol = new_mol.GetMol()\n",
    "    \n",
    "        Chem.SanitizeMol(mol)\n",
    "        mol = to_mol(to_smile(mol))\n",
    "        output = to_smile(mol)\n",
    "    except:\n",
    "        output = None\n",
    "\n",
    "    return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "def add_atom_combi(smile, atom_types, cpus=0):\n",
    "    '''\n",
    "    add_atom_combi - creates variants of `smile` with one atom \n",
    "    added or removed, defined by `atom_types`\n",
    "    \n",
    "    Inputs:\n",
    "    \n",
    "    - `smile str`: smiles string to modify\n",
    "    \n",
    "    - `atom_types list[str, int]`: list of allowed atom types to add. \n",
    "    If `-1` is in the list, variants of `smile` with one atom removed \n",
    "    will be generated. If `-2` is in the list, the code will look for \n",
    "    atoms with two neighbors, remove the center atom and bond the \n",
    "    neighbors (ie ring contraction)\n",
    "    \n",
    "    - `cpus Optional[int]`: number of cpus to use for multiprocessing. \n",
    "    If None, serial processing is used\n",
    "    '''\n",
    "    \n",
    "    bond_to_valence = {\n",
    "        Chem.rdchem.BondType.SINGLE:1,\n",
    "        Chem.rdchem.BondType.DOUBLE:2,\n",
    "        Chem.rdchem.BondType.TRIPLE:3\n",
    "    }\n",
    "    \n",
    "    mol = to_mol(smile)\n",
    "    periodic_table = Chem.GetPeriodicTable()\n",
    "    \n",
    "    valid_idxs = []\n",
    "    for atom in mol.GetAtoms():\n",
    "        imp_h = atom.GetNumImplicitHs()\n",
    "        if imp_h>0:\n",
    "            valid_idxs.append((atom.GetIdx(), imp_h))\n",
    "            \n",
    "    additions = []\n",
    "    for i, (atom_idx, imp_h) in enumerate(valid_idxs):\n",
    "        for atom_type in atom_types:\n",
    "            if (atom_type==-1) or (atom_type==-2):\n",
    "                to_add = [mol, atom_idx, atom_type, None]\n",
    "                additions.append(to_add)\n",
    "            else:\n",
    "                if atom_type in ['F', 'Cl', 'Br']:\n",
    "                    type_valence = 1\n",
    "                else:\n",
    "                    type_valence = max(list(periodic_table.GetValenceList(atom_type)))\n",
    "\n",
    "                max_valence = min(imp_h, type_valence)\n",
    "                for bt,bv in bond_to_valence.items():\n",
    "                    if bv<=max_valence:\n",
    "                        to_add = [mol, atom_idx, atom_type, bt]\n",
    "                        additions.append(to_add)\n",
    "    \n",
    "    return maybe_parallel(add_one_atom, additions, cpus=cpus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out = add_atom_combi('C1CN=CCC1', ['C', 'N', 'O', 'F', -1, -2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "def add_bond_combi(smile, max_ring_size=8, cpus=0):\n",
    "    '''\n",
    "    add_bond_combi - creates variants of `smile` with a single \n",
    "    bond added or removed\n",
    "    \n",
    "    Inputs:\n",
    "    \n",
    "    - `smile str`: smiles string to modify\n",
    "    \n",
    "    - `max_ring_size`: maximum allowed ring size\n",
    "    \n",
    "    - `cpus Optional[int]`: number of cpus to use for multiprocessing. \n",
    "    If None, serial processing is used\n",
    "    '''\n",
    "    \n",
    "    bond_to_valence = {\n",
    "        Chem.rdchem.BondType.SINGLE:1,\n",
    "        Chem.rdchem.BondType.DOUBLE:2,\n",
    "        Chem.rdchem.BondType.TRIPLE:3,\n",
    "        None:0\n",
    "    }\n",
    "    \n",
    "    allowed_ring_sizes = set([5,6,7,8])\n",
    "    \n",
    "    bond_order = [\n",
    "        None, \n",
    "        Chem.rdchem.BondType.SINGLE,\n",
    "        Chem.rdchem.BondType.DOUBLE,\n",
    "        Chem.rdchem.BondType.TRIPLE\n",
    "    ]\n",
    "    \n",
    "    bt_to_str = {Chem.rdchem.BondType.SINGLE:'single',\n",
    "                  Chem.rdchem.BondType.DOUBLE:'double',\n",
    "                  Chem.rdchem.BondType.TRIPLE:'triple',\n",
    "                  None:None}\n",
    "    \n",
    "    mol = to_mol(smile)\n",
    "    \n",
    "    valid_idxs = []\n",
    "    for atom in mol.GetAtoms():\n",
    "        imp_h = atom.GetNumImplicitHs()\n",
    "        if imp_h>0:\n",
    "            if atom.GetSymbol()=='N':\n",
    "                imp_h += 1\n",
    "            valid_idxs.append((atom.GetIdx(), imp_h))\n",
    "            \n",
    "    additions = []\n",
    "    \n",
    "    combos = list(itertools.combinations(valid_idxs,2))\n",
    "    \n",
    "    for c in combos:\n",
    "        atom1, atom2 = c\n",
    "        idx1, h1 = atom1\n",
    "        idx2, h2 = atom2\n",
    "        max_valence = min(h1, h2)\n",
    "        \n",
    "        current_bond = mol.GetBondBetweenAtoms(idx1, idx2)\n",
    "        \n",
    "        if current_bond is None:\n",
    "            path_check = len(Chem.rdmolops.GetShortestPath(mol, idx1, idx2)) in allowed_ring_sizes\n",
    "            ring_check = not (mol.GetAtomWithIdx(idx1).IsInRing() and mol.GetAtomWithIdx(idx2).IsInRing())\n",
    "            go_ahead = path_check and ring_check\n",
    "        else:\n",
    "            type_check = not (type(current_bond)==Chem.rdchem.BondType.AROMATIC)\n",
    "            go_ahead = type_check\n",
    "            \n",
    "        if go_ahead:\n",
    "#             current_bond = mol.GetBondBetweenAtoms(idx1, idx2)\n",
    "            if current_bond is None:\n",
    "                new_bonds = bond_order[1:]\n",
    "            else:\n",
    "                new_bonds = [i for i in bond_order if not i==current_bond.GetBondType()]\n",
    "            for bt in new_bonds:\n",
    "                if bond_to_valence[bt]<=max_valence:\n",
    "                    to_add = [smile, idx1, idx2, bt_to_str[bt]]\n",
    "                    additions.append(to_add)\n",
    "                \n",
    "    return maybe_parallel(add_one_bond, additions, cpus=cpus)\n",
    "\n",
    "\n",
    "def add_one_bond(inputs):\n",
    "    \n",
    "    smile, idx1, idx2, bt = inputs\n",
    "    mol = to_mol(smile)\n",
    "    \n",
    "    bond_types = {'single':Chem.rdchem.BondType.SINGLE,\n",
    "                  'double':Chem.rdchem.BondType.DOUBLE,\n",
    "                  'triple':Chem.rdchem.BondType.TRIPLE}\n",
    "    \n",
    "    new_mol = Chem.RWMol(mol)\n",
    "    current_bond = mol.GetBondBetweenAtoms(idx1, idx2)\n",
    "    \n",
    "    if bt is None:\n",
    "        new_mol.RemoveBond(idx1, idx2)\n",
    "    elif current_bond is None:\n",
    "        new_mol.AddBond(idx1, idx2, bond_types[bt])\n",
    "    else:\n",
    "        bond_idx = current_bond.GetIdx()\n",
    "        current_bond.SetBondType(bond_types[bt])\n",
    "        new_mol.ReplaceBond(bond_idx, current_bond)\n",
    "    \n",
    "    mol = new_mol.GetMol()\n",
    "    \n",
    "    try:\n",
    "        Chem.SanitizeMol(mol)\n",
    "        mol = to_mol(to_smile(mol))\n",
    "        output = to_smile(mol)\n",
    "    except:\n",
    "        output = None\n",
    "\n",
    "    return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out = add_bond_combi('C1CN=CCC1')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Proteins\n",
    "\n",
    "Functions designed for manipulating proteins as amino acid sequences.\n",
    "\n",
    "#### Current Limitations\n",
    "\n",
    "The underlying RDKit utils for amino acids are somewhat more restricted than those for SMILES strings. Only standard amino acid characters can be used (ie no wildcards). \n",
    "\n",
    "Proteins are represented as FASTA sequences, ie `MKDCSNGCSAECTGEGG`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "def to_protein(sequence_or_mol):\n",
    "    '''\n",
    "    Convert amino acid sequence to Chem.Mol\n",
    "    '''\n",
    "    if (type(sequence_or_mol) == str) or (type(sequence_or_mol) == np.str_):\n",
    "        mol = Chem.MolFromFASTA(sequence_or_mol)\n",
    "        if mol is not None:\n",
    "            try:\n",
    "                Chem.SanitizeMol(mol)\n",
    "            except:\n",
    "                mol = None\n",
    "    else:\n",
    "        mol = sequence_or_mol\n",
    "        \n",
    "    return mol\n",
    "\n",
    "def to_sequence(sequence_or_mol):\n",
    "    '''\n",
    "    Converts sequence Mol into string\n",
    "    '''\n",
    "    if type(sequence_or_mol)==Chem.Mol:\n",
    "        sequence = Chem.MolToSequence(sequence_or_mol)\n",
    "    else:\n",
    "        sequence = sequence_or_mol\n",
    "        \n",
    "    return sequence\n",
    "\n",
    "def to_proteins(list_of_inputs):\n",
    "    return maybe_parallel(to_protein, list_of_inputs)\n",
    "\n",
    "def to_sequences(list_of_inputs):\n",
    "    return maybe_parallel(to_sequence, list_of_inputs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert type(to_protein('MKDCSNGCSAECTGEGG'))==Chem.Mol\n",
    "assert to_sequence(to_protein('MKDCSNGCSAECTGEGG')) == 'MKDCSNGCSAECTGEGG'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted 00_core.ipynb.\n",
      "Converted 01_chem.ipynb.\n",
      "Converted 02_template.filters.ipynb.\n",
      "Converted 03_template.template.ipynb.\n",
      "Converted 04_template.blocks.ipynb.\n",
      "Converted index.ipynb.\n",
      "Converted tutorials.ipynb.\n",
      "Converted tutorials.structure_enumeration.ipynb.\n"
     ]
    }
   ],
   "source": [
    "# hide\n",
    "from nbdev.export import notebook2script; notebook2script()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (mrl)",
   "language": "python",
   "name": "mrl"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
